// zoho-extension-toolkit/lib/authentication/login.js
'use strict';

const portfinder = require('portfinder');
const path = require('path');
const open = require('open');
const font = require('ansi-colors');

const _ = require('../util_modules');
const API = require('../internal/api');
const Credential = require('../internal/credential');
const store = require('../configstore');

class Login {
	constructor(localhost = true, user = true) {
		this.localhost = localhost;
		this.portPromise = portfinder.getPortPromise({
			startPort: 9005,
			port: 9005,
			stopPort: 9009
		});
		this.salt = localhost ? 'w' : 'm';
		this.user = user;
	}

	init() {
		if (!this.localhost) {
			return this._loginWithoutLocalhost();
		}
		return this.portPromise.then((port) => this._loginWithLocalhost(port));
	}

	_getCallbackUrl(port) {
		return 'http://localhost:' + port;
	}

	_getLoginUrl(callbackUrl) {
		return (
			_.CONSTANTS.ORIGIN.admin +
			'/oauth/v2/auth?' +
			_.JS.map(
				{
					client_id: _.CONSTANTS.AUTH.web.id,
					scope: _.JS.values(_.CONSTANTS.SCOPE).join(' '),
					response_type: 'code',
					access_type: 'offline',
					prompt: 'consent',
					redirect_uri: callbackUrl
				},
				(v, k) => {
					return k + '=' + encodeURIComponent(v);
				}
			).join('&')
		);
	}

	_getTokenFromAuthorizationCode(code, callbackUrl) {
		return new API()
			.post('/oauth/v2/token', {
				origin: _.CONSTANTS.ORIGIN.auth,
				dctype: store.get('dctype'),
				params: {
					code: code,
					client_id: _.CONSTANTS.AUTH.web.id,
					client_secret: _.CONSTANTS.AUTH.web.secret,
					redirect_uri: callbackUrl,
					grant_type: 'authorization_code'
				},
				auth: false,
				log: {
					skipQuery : true 
				}
			})
			.then((res) => {
				if (
					!_.JS.hasIn(res, 'body.access_token') ||
					!_.JS.hasIn(res, 'body.refresh_token')
				) {
					_.LOGGER.debug(
						'Token Fetch Error: \n' + ' Status: ' + res.status + ', Body: ' + res.body
					);
					_.LOGGER.error(
						'Unable to get refresh_token from code. Kindly try again. \n'
					);
					process.exit(1);
				}
				let token = _.JS.assign(
					{
						created_time: Date.now(),
						expires_at: Date.now() + res.body.expires_in
					},
					res.body
				);
				return token;
			});
	}

	_getTokenFromDeviceCode(code, retryCount = 0) {
		return new API()
			.post('/oauth/v3/device/token', {
				origin: _.CONSTANTS.ORIGIN.auth,
				dctype: store.get('dctype'),
				params: {
					client_id: _.CONSTANTS.AUTH.mobile.id,
					client_secret: _.CONSTANTS.AUTH.mobile.secret,
					scope: _.JS.values(_.CONSTANTS.SCOPE).join(' '),
					grant_type: 'device_token',
					code
				},
				log: {
					skipQuery: true
				},
				auth: false
			})
			.then((res) => {
				if (
					!_.JS.hasIn(res, 'body.access_token') ||
					!_.JS.hasIn(res, 'body.refresh_token')
				) {
					_.LOGGER.debug(
						'Token Fetch Error: \n' +
							' Status: ' +
							res.status +
							', Body: ' +
							JSON.stringify(res.body)
					);
					_.LOGGER.debug('> polling <');
					if (++retryCount > 10) {
						_.LOGGER.error(
							'Unable to get refresh_token from code. Kindly try again. \n' +
								'If the problem persists contact support@zohosigma.com'
						);
						process.exit(1);
					}
					return _.JS.sleep(2000).then(() => {
						return this._getTokenFromDeviceCode(code);
					});
				}
				let token = _.JS.assign(
					{
						created_time: Date.now(),
						expires_at: Date.now() + res.body.expires_in
					},
					res.body
				);
				return token;
			})
			.catch((err) => {
				_.LOGGER.debug('Token Fetch Error: \n' + err);
				_.LOGGER.error(err.message);
				process.exit(1);
			});
	}

	_respondWithFile(req, res, statusCode, filename) {
		return _.FS.ASYNC.readFile(path.join(__dirname, filename)).then((content) => {
			res.writeHead(statusCode, {
				'Content-Length': content.length,
				'Content-Type': 'text/html'
			});
			res.end(content);
			req.socket.destroy();
		});
	}

	_getUserDetails() {
		if (!this.user) {
			return '';
		}
		return new API()
			.get('/oauth/user/info', {
				origin: _.CONSTANTS.ORIGIN.auth,
				dctype: store.get('dctype'),
				auth: true
			})
			.then((res) => {
				if (!_.JS.has(res, 'body.Email') && !_.JS.has(res, 'body.ZUID')) {
					_.LOGGER.debug(
						'User Details Fetch Error:\n' +
							' Status: ' +
							res.statusCode +
							', Body: ' +
							res.body
					);
					return null;
				}
				return res.body;
			});
	}

	_loginWithoutLocalhost(err) {
		if (err) {
			_.LOGGER.debug('Error while login with server : ' + err.message);
		}
		this.salt = 'm';
		let token = {};
		const throbber = new (require('../spinner'))().getThrobber();

		return new API()
			.post('/oauth/v3/device/code', {
				origin: _.CONSTANTS.ORIGIN.auth,
				dctype: store.get('dctype'),
				params: {
					client_id: _.CONSTANTS.AUTH.mobile.id,
					scope: _.JS.values(_.CONSTANTS.SCOPE).join(' '),
					grant_type: 'device_request',
					access_type: 'offline',
					prompt: 'consent'
				},
				auth: false
			})
			.then((res) => {
				if (!_.JS.hasIn(res, 'body.user_code') || !_.JS.hasIn(res, 'body.device_code')) {
					_.LOGGER.debug(
						'Token Fetch Error: \n' +
							' Status: ' +
							res.status +
							', Body: ' +
							JSON.stringify(res.body)
					);
					_.LOGGER.error(
						'Unable to get code from zoho. Kindly check your credentials and try again. \n' +
							'If the problem persists contact support@zohosigma.com'
					);
					process.exit(1);
				}
				_.LOGGER.info();
				_.LOGGER.info('Visit this URL on any device: \n');
				_.LOGGER.info(font.bold.underline(_.JS.get(res, 'body.verification_url')));
				_.LOGGER.info();
				_.LOGGER.info(
					'Enter the device verification code: ' +
						font.bold(_.JS.get(res, 'body.user_code')) +
						' and click verify to continue.'
				);
				return _.JS.sleep(10000).then(() => {
					return _.JS.get(res, 'body.device_code');
				});
			})
			.then((deviceCode) => {
				throbber.start('checking if the code has been enterted');
				return this._getTokenFromDeviceCode(deviceCode);
			})
			.then((result) => {
				token = result;
				_.JS.set(token, 'token', this.salt + '_' + result.refresh_token);
				Credential._oneTimeToken = token.access_token;
				throbber.text = 'getting user details';
				return this._getUserDetails();
			})
			.then((details) => {
				throbber.stop();
				return {
					user: details,
					token,
					scopes: _.JS.values(_.CONSTANTS.SCOPE)
				};
			})
			.catch((err) => {
				throbber.stop();
				return Promise.reject(err);
			});
	}

	_loginWithLocalhost(port) {
		let self = this;
		return new Promise((resolve, reject) => {
			let callbackUrl = self._getCallbackUrl(port);
			let authUrl = self._getLoginUrl(callbackUrl);
			let reqCount = 0;
			let server = require('http').createServer((req, res) => {
				let processReq = true;
				if (reqCount + 1 > 1 || req.url === '/favicon.ico') {
					_.LOGGER.debug('unknown request received : ' + req.url);
					processReq = false;
					res.writeHead(404);
					res.end();
				}
				reqCount += 1;
				let token = {};
				let query = _.JS.get(require('url').parse(req.url, true), 'query', {});
				if (_.JS.has(query, 'code')) {
					store.set('dctype',query.location);
					_.LOGGER.debug('DCType Of the user is '+store.get('dctype'));
					return this._getTokenFromAuthorizationCode(query.code, callbackUrl)
						.then((result) => {
							token = result;
							_.JS.set(token, 'token', self.salt + '_' + result.refresh_token);
							Credential._oneTimeToken = token.access_token;
							return self._respondWithFile(
								req,
								res,
								200,
								'../../templates/loginSuccess.html'
							);
						})
						.then(() => {
							return self._getUserDetails();
						})
						.then((details) => {
							server.close();
							return resolve({
								user: details,
								token,
								scopes: _.JS.values(_.CONSTANTS.SCOPE)
							});
						})
						.catch((err) => {
							server.close();
							_.LOGGER.error(err);
							process.exit(1);
						});
				} else if (processReq) {
					return self
						._respondWithFile(req, res, 400, '../../templates/loginFailure.html')
						.then(() => {
							server.close();
							_.LOGGER.error("Credentials doesn't seem to be valid.\n");
							process.exit(1);
						});
				}
			});

			server.listen(port, () => {
				_.LOGGER.info();
				_.LOGGER.info('Click this URL to login to your Zoho Account');
				_.LOGGER.info(font.bold.underline(authUrl));
				_.LOGGER.info('Note: To login to CN DC, update the above URL from accounts.zoho.com to accounts.zoho.com.cn');
				_.LOGGER.info();
				open(authUrl);
			});

			server.on('error', () => {
				self._loginWithoutLocalhost().then(resolve, reject);
			});
		});
	}
}

module.exports = Login;
