// zoho-extension-toolkit/lib/util_modules/lib/logger.js
'use strict';

const font = require('ansi-colors');
const char = require('./char');
const store = require('../../configstore');
const logger = require('../../logger');


const self = (module.exports = {
	log: (level, message) => {
		self.disableLog(store.get('usage'));
		level = level || 'info';
		logger[level](message);
	},
	info: (message = '') => {
		self.disableLog(store.get('usage'));
		logger['info'](message);
	},
	debug: (message = '') => {
		self.disableLog(store.get('usage'));
		logger['debug'](message);
	},
	success: (message, level = 'info', label = '') => {
		let postChar = label === '' ? '' : label + ': ';
		self.log(level, font.green(char.success + ' ' + postChar) + message);
	},
	message: (message, level = 'info', label = '') => {
		let postChar = label === '' ? '' : label + ': ';
		self.log(level, font.cyan(char.info + ' ' + postChar) + message);
	},
	warning: (message, level = 'warn', label = '') => {
		let postChar = label === '' ? '' : label + ': ';
		self.log(level, font.yellow(char.warning + ' ' + postChar) + message);
	},
	error: (message, level = 'error', label = '') => {
		let postChar = label === '' ? '' : label + ': ';
		self.log(level, font.red(char.error + ' ' + postChar) + message);
	},
	labeled: (label, message, level = 'info') => {
		return {
			SUCCESS: () => self.success(message, level, label),
			WARN: () => self.warning(message, level, label),
			MESSAGE: () => self.message(message, level, label),
			ERROR: () => self.error(message, level, label)
		};
	},
	disableLog: (flag=true) =>{
		if(!flag){
			for(var i=0;i<logger.transports.length;i++){
				if(logger.transports[i].name == 'FileLog'){
					logger.transports[i].silent = true;
					break;
				}
			}
		}	
	}
});
