// zoho-extension-toolkit/lib/util_modules/lib/option.js
'use strict';

const _ = require('./js');

module.exports = {
	getInheritedOption: (options) => (key, fallback = null) => {
		let target = options;
		while (target) {
			if (_.has(target, key)) {
				return target[key];
			}
			target = target.parent;
		}
		return fallback;
	},

	properties: (options) => {
		return {
			set: (key, value) => {
				_.set(options, key, value);
			},
			get: (key, fallback = null) => {
				return _.get(options, key, fallback);
			}
		};
	},

	context: (options) => {
		return {
			set: (key, value) => {
				_.set(options, '__context.' + key, value);
			},
			get: (key, fallback = null) => {
				return _.get(options, '__context.' + key, fallback);
			}
		};
	},

	payload: (options) => {
		return {
			set: (key, value) => {
				_.set(options, '__payload.' + key, value);
			},
			get: (key, fallback = null) => {
				return _.get(options, '__payload.' + key, fallback);
			}
		};
	},

	getCurrentCommand: (options, fallback = null) => {
		return _.get(options, '_name', fallback);
	},

	getCommandOption: (options) => (key, fallback = null) => {
		return _.get(options, key, fallback);
	},

	setGlobalOption: (options) => (key, value) => {
		_.set(options, '__custom.globalOpts.' + key, value);
	},

	getGlobalOption: (options) => (key, fallback = null) => {
		return _.get(options, '__custom.globalOpts.' + key, fallback);
	}
};
