// zoho-extension-toolkit/lib/util_modules/lib/fs/lib/sync.js
'use strict';

const fs = require('fs-extra');
const os = require('os');
const path = require('path');

module.exports = {
	getWriteStream: (pth) => {
		return fs.createWriteStream(pth);
	},

	getReadStream: (pth) => {
		return fs.createReadStream(pth);
	},

	fileExists: (path) => {
		try {
			let stats = fs.statSync(path);
			return stats.isFile();
		} catch (e) {
			return false;
		}
	},

	dirExists: (path) => {
		try {
			let stats = fs.statSync(path);
			return stats.isDirectory();
		} catch (e) {
			return false;
		}
	},

	getAllDirs: (srcPath) => {
		let directories = fs
			.readdirSync(srcPath)
			.map((name) => {
				return path.join(srcPath, name);
			})
			.filter((source) => {
				return fs.lstatSync(source).isDirectory();
			});
		return directories;
	},

	deleteFile: (pth) => {
		return fs.removeSync(pth);
	},

	readFile: (path, encoding = 'utf8') => {
		if (module.exports.fileExists(path)) {
			return fs.readFileSync(path, encoding);
		}
		return null;
	},

	ensureDir: (dir) => {
		return fs.ensureDirSync(dir);
	},

	ensureFile: (filePth) => {
		return fs.ensureFileSync(filePth);
	},

	writeFile: (path, data, encoding = 'utf8') => {
		return fs.writeFileSync(path, data, { encoding });
	},

	appendFile: (path, data, encoding = 'utf8') => {
		return fs.appendFileSync(path, data, { encoding });
	},

	modifyFileName: (pth, modifier) => {
		if (Array.isArray(pth)) {
			return pth.map((el) => module.exports.modifyFileName(el, modifier));
		}

		const ext = path.extname(pth);
		return path.join(path.dirname(pth), modifier(path.basename(pth, ext), ext));
	},

	renameFile: (pth, modifier) => {
		if (Array.isArray(pth)) {
			return pth.map((el) => module.exports.renameFile(el, modifier));
		}
		let modifiedName = module.exports.modifyFileName(pth, modifier);
		fs.renameSync(pth, modifiedName);
		return modifiedName;
	},

	isPathOutside: (from, to) => {
		const relative = path.relative(from, to);
		return !(relative && !relative.startsWith('..') && !path.isAbsolute(relative));
	},

	deleteDir: (dir) => {
		return fs.removeSync(dir);
	},

	deleteTempDir: () => {
		let dir = path.join(os.tmpdir(), '.catalyst');
		return module.exports.deleteDir(dir);
	},

	copyFile: (src, dest) => {
		module.exports.ensureDir(path.dirname(dest));
		return fs.copyFileSync(src, dest);
	},

	copyDir: (srcDir, destDir) => {
		return fs.copySync(srcDir, destDir);
	}
};
