// zoho-extension-toolkit/lib/utils/service.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/

var rulesConfig = require('./../rules/rulesconfig.js');
var jsonfile = require('jsonfile');
var path = require('path');
var CONST = require('../commands/constants.js');
const _ = require('../util_modules');
var chalk = require('chalk');
const fs = require('fs');
const crypt = new (require('../internal/crypt'))('ZC_TRAM');
const store = require('../../lib/configstore');
var i18next = require('i18next');
var util = require('../utils');
var isObject = util.isObject;
var isArray = util.isArray;

var SERVICE_NAMES = [
  {
    name: 'Zoho CRM',
    value: CONST.SERVICE.CRM
  },
  {
    name: 'Catalyst',
    value: CONST.SERVICE.CATALYST
  },
  {
    name: 'Zoho Projects',
    value: CONST.SERVICE.PROJECTS
  },
  {
    name: 'Zoho Desk',
    value: CONST.SERVICE.DESK
  },
  {
    name: 'Qntrl',
    value: CONST.SERVICE.QNTRL
  },
  {
    name: 'Zoho Mail',
    value: CONST.SERVICE.MAIL
  },
  {
    name: 'Zoho Creator',
    value: CONST.SERVICE.CREATOR
  },
  {
    name: 'Zoho TeamInbox',
    value: CONST.SERVICE.TEAMINBOX
  },
  {
    name: 'ServiceDesk Plus Cloud',
    value: CONST.SERVICE.SDP
  },
  {
    name: 'Zoho IOT',
    value: CONST.SERVICE.IOT
  },
  {
    name: 'Zoho Books',
    value: CONST.SERVICE.FINANCE
  },
  {
    name : 'Zoho Expense',
    value : CONST.SERVICE.EXPENSE
  },
  {
    name : 'Zoho Inventory',
    value : CONST.SERVICE.INVENTORY
  },
  {
    name : 'Zoho People',
    value : CONST.SERVICE.PEOPLE
  },
  {
    name : 'Bigin',
    value : CONST.SERVICE.BIGIN
  },
  {
    name : 'Zoho Billing',
    value : CONST.SERVICE.BILLING
  },
  {
    name : 'Zoho Recruit',
    value : CONST.SERVICE.RECRUIT
  },
  {
    name : 'Zoho FSM',
    value : CONST.SERVICE.FSM
  },
  {
    name : 'Log360 Cloud',
    value : CONST.SERVICE.LOGS360CLOUD
  }

];

var CSP_DOMAINS_ARR = ["connect-src"];
let cspArrayLength = 10;

var ZET_API_MODEL_SERVICE_NAMES = [];
var DEFAULT_TEMPLATE = 'default';
function getServiceName(projectRootDir) {
  var manifestFile = path.join(projectRootDir, CONST.PLUGIN_MANIFEST);
  var manifestData = require(manifestFile);

  return manifestData.service;
}
function setServiceName(projectRootDir, serviceName, cb) {
  var manifestFile = path.join(projectRootDir, CONST.PLUGIN_MANIFEST);
  var manifestData = require(manifestFile);
  manifestData["service"] = serviceName;
  jsonfile.writeFile(manifestFile, manifestData, { spaces: 2 }, function (err) {
    if (err) {
      console.error(err);
    }
  });
}
function getServiceNames() {
  return SERVICE_NAMES;
}
function getValidationRules(projectRootDir) {
  var ruleConfig = {};
  try {
    var serviceName = getServiceName(projectRootDir);
    ruleConfig = rulesConfig[serviceName] || rulesConfig['DEFAULT'];
  } catch (err) {
    ruleConfig = rulesConfig['DEFAULT'];
  }

  return ruleConfig;
}
function getServiceTemplateName(serviceName) {
  return CONST.SERVICE[serviceName];
}
function getDefaultTemplateName() {
  return DEFAULT_TEMPLATE;
}

function isValidServiceName(serviceName) {
  var _filteredList = getServiceNames().filter(function (service) {
    return service.value === serviceName;
  });

  if (Array.isArray(_filteredList) && _filteredList.length === 1) return true;
}
function getPrePackRule(projectRootDir) {
  var ruleConfig = getValidationRules(projectRootDir);
  return ruleConfig.PRE_PACK;
}
function getPostPackRule(projectRootDir) {
  var ruleConfig = getValidationRules(projectRootDir);
  return ruleConfig.POST_PACK;
}
function isZetAPIModelService(projectRootDir) {
  var manifestFile = path.join(projectRootDir, CONST.PLUGIN_MANIFEST);
  var manifestData = require(manifestFile);
  var serviceNames = getServiceNames().map(function (obj) { return obj.value; });
    if (serviceNames.indexOf(manifestData.service) === -1) {
      _.LOGGER.info(chalk.red('Invalid service. Allowed services are : ')+chalk.bold.red(serviceNames.join(', ')));
      process.exit(1);
    }
  if(CONST.ZET_API_MODEL_SERVICES.indexOf(manifestData.service) != -1){
    return true
  }
  return false;
}

function writeJsonAsFile(json,projectRootDir){
  return new Promise(function (resolve, reject) {
    var json_path = path.join(projectRootDir, CONST.APP_JSON);
    fs.writeFile(json_path , crypt.encrypt(json), (err) => {
        if (err) {
          _.LOGGER.debug('Error in creating the json file:'+err);
        };
          _.LOGGER.debug('Data written in app_info.json file');
          resolve();
    }); 
  });
}

function isJsonFilePresent(projectRootDir){
  var json_path = path.join(projectRootDir, CONST.APP_JSON);
  if (fs. existsSync(json_path)) { 
    _.LOGGER.debug("app_info.json File exists.");
    return true;
  } else { 
    _.LOGGER.debug("app_info.json File does not exist.");
    return false;
  } 
}

function readJsonFromFile(projectRootDir){
  return new Promise(function (resolve, reject) {
  var json_path = path.join(projectRootDir, CONST.APP_JSON);
  fs.readFile(json_path, (err, data) => {
    if (err) {
      _.LOGGER.debug('Error in reading the json file:'+err);
    }
    var buf = data.toString("utf-8")
    var jsondata = crypt.decrypt(buf);
    resolve(jsondata);
  });
  });
}

function isZetAPIModelServiceByServiceName(serviceName) {
  if(CONST.ZET_API_MODEL_SERVICES.indexOf(serviceName) != -1){
    return true
  }
  return false;
}

function getZetAPIModelServiceNames() {
  SERVICE_NAMES.forEach(function (arrayItem) {
    if(CONST.ZET_API_MODEL_SERVICES.indexOf(arrayItem.value ) != -1){
       ZET_API_MODEL_SERVICE_NAMES.push(arrayItem);
    }
});
  return ZET_API_MODEL_SERVICE_NAMES;
}

function validateTokenObject(tokenObj) {
  if(tokenObj == null){
    _.LOGGER.error(chalk.red(i18next.t("login_error")));
    process.exit(1);
  }
}

function validateDefaultWorkspace(workspace) {
  if(!workspace){
    _.LOGGER.error(chalk.red('Workspace not found. Use command \'ZET list_workspace\' to find Workspace. '));
    process.exit(1);
  }
}

function validateAppInfoFile(isFilePresent) {
  if(!isFilePresent){
    _.LOGGER.error(chalk.red(i18next.t("info_error")));
    process.exit(1);
  }
}

function validateJSONData(app_data,workspace) {
  validateDefaultWorkspace(workspace);
  if(app_data.workspaceName != workspace){
    _.LOGGER.error(chalk.red('The Extension belong to \''+app_data.workspaceName+'\' workspace. Kindly use ')+ chalk.bold.red('Zet list_workspace')+chalk.red(' to switch workspace'));
    process.exit(1);
  }
  if(!app_data.appUuid){
    _.LOGGER.error(chalk.red(i18next.t("info_error")));
    process.exit(1);
  }
}

 function getPath(object){
  var path = object.name1 ? object.name1 : object.name2;
  if(object.relativePath.includes('app')){
    path = object.relativePath+'/'+path;
  }
  return path;
}

function sessionExpireCheck(){
  return new Promise(function (resolve, reject) {
    var deathLine = store.get('oAuthCreatedTime') + (CONST.ExpireNoOfDays * 24 * 60 * 60 * 1000) ;
      if( store.get('oAuthCreatedTime') && deathLine < Date.now()){
        _.LOGGER.debug('Session Expired');
        require('../authentication').autoLogout().then(()=>{
          resolve()
        });
      }else{
        resolve();
      }
  });
}

function validateCspDomains(cspDomains){
  var errors = [];
  if (!isObject(cspDomains) || cspDomains.constructor !== Object) {
    errors.push(i18next.t('pluginmanifest_invalidCSPDomains'));
  } else {             
    CSP_DOMAINS_ARR.every((element,i) => {
        if(element) {                      
            if(isArray(cspDomains[element])) {
                if(cspDomains[element].length > cspArrayLength) {
                    errors.push(element + i18next.t('pluginmanifest_csp_size_exceeded') + cspArrayLength);
                }
                if(util.isDuplicateURL(cspDomains[element])) {
                    errors.push(i18next.t('pluginmanifest_csp_duplicatedomain') + element);
                } 
                if(!util.isValidCSPDomain(cspDomains[element])) {
                    errors.push(cspDomains[element] + i18next.t('pluginmanifest_invalidConnectElementEntry') + element);
                }   
            }
            else {
                errors.push(element + i18next.t('pluginmanifest_invalidConnectArray'));
                return "";
            }
        }
    });                   
  }
  return errors;
}

module.exports = {
  getValidationRules: getValidationRules,
  getServiceTemplateName: getServiceTemplateName,
  getServiceName: getServiceName,
  setServiceName: setServiceName,
  getServiceNames: getServiceNames,
  isValidServiceName: isValidServiceName,
  getPrePackRule: getPrePackRule,
  getPostPackRule: getPostPackRule,
  isZetAPIModelService: isZetAPIModelService,
  writeJsonAsFile: writeJsonAsFile,
  isJsonFilePresent: isJsonFilePresent,
  readJsonFromFile: readJsonFromFile,
  isZetAPIModelServiceByServiceName: isZetAPIModelServiceByServiceName,
  getZetAPIModelServiceNames : getZetAPIModelServiceNames,
  validateTokenObject : validateTokenObject,
  validateDefaultWorkspace : validateDefaultWorkspace,
  validateAppInfoFile : validateAppInfoFile,
  validateJSONData : validateJSONData,
  getPath : getPath,
  sessionExpireCheck : sessionExpireCheck,
  validateCspDomains : validateCspDomains
};
