// zoho-extension-toolkit/lib/rules/DESK.PluginManifestRule.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/

var path = require('path');
var fs = require('fs');
var i18next = require('i18next');
var util = require('../utils');
var CONST = require('../commands/constants');
var service = require('./../utils/service.js');
var isUndefined = util.isUndefined;
var isBoolean = util.isBoolean;
var isObject = util.isObject;
var isString = util.isString;
var isArray = util.isArray;
var isAllowedString = util.isAllowedString;
var isValidURL = util.isValidURL;

var LOCATION_CONSTANT = [
    "desk.ticket.detail.rightpanel", 
    "desk.ticket.detail.subtab", "desk.bottomband", "desk.topband", "desk.background",
    "desk.contact.detail.rightpanel", "desk.contact.detail.subtab", "desk.extension.preference", 
    "desk.account.detail.rightpanel", "desk.account.detail.subtab", "desk.ticket.detail.lefttab", 
    "desk.contact.detail.lefttab", "desk.account.detail.lefttab", "desk.ticket.detail.moreaction", 
    "desk.ticket.thread.moreaction", "desk.ticket.form.rightpanel", "desk.ticket.form.sidepanel", 
    "desk.contact.form.rightpanel", "desk.contact.form.sidepanel", "desk.account.form.rightpanel", 
    "desk.account.form.sidepanel", "helpcenter.ticket.detail", "desk.extension.telephony",
    "helpcenter.ticket.form", "helpcenter.ticket.list", 
    "helpcenter.invisible", "desk.invisible", "helpcenter.homepage", "helpcenter.kb.homepage", 
    "helpcenter.kb.article.list","helpcenter.kb.article.detail", "helpcenter.community.homepage", 
    "helpcenter.community.topic.list", "helpcenter.community.topic.detail","desk.kb.article.detail.subtab",
    "desk.kb.managekb.lefttab","desk.ticket.blueprint.transition"
];
var LOCALE_CODE = [
    "en", "ja", "zh", "tw", "es", "de", "fr", "tr", "ru", "pt", "it", "nl", "da", "pl",
    "sv", "id", "hi", "ro", "ta", "te", "he", "ar"
];
var MANIFEST_KEY = [
    "locale", "storage", "service", "modules", "connectors", 
    "config", "whiteListedDomains"
];

var CSP_DOMAINS_ARR = ["connect-src", "img-src", "style-src", "script-src"];
let cspArrayLength = 10;

function validatePluginManifest(projectRootDir) {
    var errors = [];
    var manifest = require(path.join(projectRootDir, CONST.PLUGIN_MANIFEST));
    let manifestKeys = Object.keys(manifest);

    //plugin manifest mandatory key validation
    for (var i = 0; i < MANIFEST_KEY.length; i++) {
        if (manifestKeys.indexOf(MANIFEST_KEY[i]) === -1) {
            errors.push("\"" + MANIFEST_KEY[i] + "\"" + i18next.t('pluginmanifest_keymissing'));
        }
    }
    if( errors.length > 0 ) return errors;

    var defaultLocale = "en";
    var isDefaultLocalePresent = false;
    var isSingleInvalidCode = false ;
    var errorLocaleCodes = "" ;
    var hasLocaleError = false;
    //locale file exists validation
    if (manifest.locale) {
        if( isArray(manifest.locale) ) {
            var locales = manifest.locale;
            for( var i=0;i<locales.length; i++) {
                let _locale = locales[i];
                let relativePath = path.join( 'app', 'translations', _locale + '.json');
                let localePath = path.join(projectRootDir, relativePath);
                if(defaultLocale === _locale){
                    isDefaultLocalePresent = true;
                }
                if(LOCALE_CODE.indexOf(_locale) === -1){
                   if(!hasLocaleError){
                    errorLocaleCodes += "'"+_locale+"'"
                    isSingleInvalidCode = true;
                   }else{
                    errorLocaleCodes += ",'"+_locale+"'"
                    isSingleInvalidCode = false;
                   }
                   hasLocaleError = true;
                    continue;
                }
                if (!fs.existsSync(localePath)) {
                    errors.push(i18next.t('desk_pluginmanifest_localenotfound') + relativePath);
                }
            }
            if (hasLocaleError) {
                var errorURL = "https://www.zoho.com/desk/extensions/guide/#plugin-manifest";
                if(isSingleInvalidCode){
                    errors.push("Invalid locale code "+errorLocaleCodes+" found in locale array. To know the valid codes visit '"+errorURL+"'.");
                }else{
                    errors.push("Invalid locale codes "+errorLocaleCodes+" found in locale array. To know the valid codes visit '"+errorURL+"'.");
                }
            }
            if(!isDefaultLocalePresent){
                errors.push("Default locale code '"+defaultLocale+"' must be available in locale array.");
            }
        } else {
            errors.push(i18next.t('desk_pluginmanifest_locale_invalidtype'));
        }
    }

    if (typeof manifest.storage !== 'boolean' ) {
        errors.push(i18next.t('desk_pluginmanifest_invalidstorage'));
    }

    if (manifest.type !== undefined) {
        if (!isString(manifest.type) || ( isString(manifest.type) && manifest.type.trim() === "" ) ||
            (manifest.type !== 'personal' && manifest.type !== 'org')) {
            errors.push(i18next.t('desk_pluginmanifest_invalidtype'));
        }
    }

    if (!isUndefined(manifest.modules)) {
        if (!isObject(manifest.modules)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidmodules'));
        } else if (isUndefined(manifest.modules.widgets)) {
            errors.push(i18next.t('desk_pluginmanifest_widgetkeymissing'));
        } else if (!isArray(manifest.modules.widgets)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidwidgettype'));
        } else if (isArray(manifest.modules.widgets) && manifest.modules.widgets.length === 0) {
            errors.push(i18next.t('desk_pluginmanifest_widgetisempty'));
        } else {
            errors = errors.concat(validateWidgetConfigs(projectRootDir, manifest.modules.widgets));
        }
    }

    if (manifest.whiteListedDomains) {
        if (!isArray(manifest.whiteListedDomains)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidwhitelisteddomaintype'));
        } else {
            if (validateDomains(manifest.whiteListedDomains)) {
                errors.push(i18next.t('desk_pluginmanifest_invalidwhitelisteddomain'));
            }
        }
    }

    // 'connectors' validation
    var connectorsConfig = manifest.connectors;
    if (!isUndefined(connectorsConfig)) {
        if (isArray(connectorsConfig)) {
            errors = errors.concat(validateConnectorConfig(connectorsConfig));
        } else {
            errors.push(i18next.t('desk_pluginmanifest_invalidconnectortype'));
        }
    }

    // 'deskOAuth' validation
    var deskOAuthConfig = manifest.deskOauth;
    if (!isUndefined(deskOAuthConfig)) {
        if (isObject(deskOAuthConfig)) {
            errors = errors.concat(validateDeskOAuthConfig(deskOAuthConfig));
        } else {
            errors.push(i18next.t('desk_manifest_doc_invalidtype'));
        }
    }

    // 'TPAOAuth' validation
    var tpaOAuth = manifest.TPAOauth;
    if( !isUndefined(tpaOAuth) ) {
        if(isArray(tpaOAuth)) {
            errors = errors.concat(validateTPAOAuthConfig(tpaOAuth));
        } else {
            errors.push(i18next.t('desk_manifest_tpo_invalidtype'));
        }
    }

    // 'callbackListeners' validation
    var callbackListener = manifest.callbackListener;
    if( !isUndefined(callbackListener) && isObject(callbackListener) ) {
        var callbackListenerKeys = [
            'onDeskAuthorise', 'onTPAAuthorise', 'onTPARevoke', 
            'onUpdate', 'onConfigParamChange', 'onUninstall' 
        ];
        for( var i=0;i<callbackListenerKeys.length;i++) {
            var cbKey = callbackListenerKeys[i];
            if( !isUndefined(callbackListener[cbKey]) && !isValidURL(callbackListener[cbKey]) ) {
                errors.push('Invalid URL value provided in manifest for key: '+ cbKey);
            }
        }
        var secret = manifest.secret;
        if(isUndefined(secret)) {
            errors.push(i18next.t('desk_pluginmanifest_secretkeymissing'));
        } else if (!isString(secret)){
            errors.push(i18next.t('desk_pluginmanifest_secretkeyinvalid'));
        }
    } else if ( !isUndefined(callbackListener) && !isObject(callbackListener) ) {
        errors.push(i18next.t('desk_pluginmanifest_callbacklistener_invalidtype'));
    }

    // 'zohoAuthorisation' Validation
    var authorisation = manifest.zohoAuthorisation;
    if(!isUndefined(authorisation)) {
       /* if(isObject(authorisation)) {
            if(!isUndefined(authorisation.type)) {
                if(authorisation.type != "oauth" && authorisation.type != "connectors") {
                    errors.push(i18next.t('desk_pluginmanifest_invalidtypevalue'));
                }
            } else {
                errors.push(i18next.t('desk_pluginmanifest_missingauthorizationtype'));
            }
        } else {
            errors.push(i18next.t('desk_pluginmanifest_invalidauthorizationtype'));
        }*/
        if(isObject(authorisation)) {
            if( Object.keys(authorisation).length !== 0){
            
                if(isUndefined(authorisation.type)){
                    errors.push(i18next.t('desk_pluginmanifest_missingauthorizationtype'));
                }
                if(!isUndefined(authorisation.serviceName)){
                    if(authorisation.serviceName!="zlabs_integration"){
                        errors.push(i18next.t('desk_pluginmanifest_authorizationinvalidserviceName'));
                    }
                }
            }
        }else {
            errors.push(i18next.t('desk_pluginmanifest_invalidauthorizationtype'));
        }
        
    }

    // 'config' validations
    var configs = manifest.config;
    if (!isUndefined(configs)) {
        if (isArray(configs)) {
            if (configs.length > 20) {
                errors.push(i18next.t('desk_pluginmanifest_configlimitexceeded'));
            } else {
                errors = errors.concat(validateConfigOptions(configs));
            }
        } else {
            errors.push(i18next.t('desk_pluginmanifest_invalidconfigtype'));
        }
    }

    // 'cspDomains' validations
    if (manifest.cspDomains) {
        let cspDomains = manifest.cspDomains;
        errors = errors.concat(service.validateCspDomains(cspDomains));
    }

    var resourcejsonpath = path.join(projectRootDir,CONST.RESOURCE_JSON);
    if(fs.existsSync(resourcejsonpath)) {
        var resourcejson = require(resourcejsonpath);
        errors = errors.concat(validateResourcesJson(resourcejson, projectRootDir));
    }else{
        errors.push(i18next.t('desk_resource_missing_jsonfile'));
    }

    return errors;
}

function validateResourcesJSONCA(customActions){
    var errors = [];
    customActions.forEach(ca=>{
        if(ca.hasOwnProperty('resourceName')){
            var resourceName = ca.resourceName || '';
            if(resourceName.length > 250){
                errors.push(i18next.t('desk_ca_reasourcename_maxlength'));
            }
        }
        /** Custom Action Name validation */
        if(!ca.hasOwnProperty('customActionName')){
            var i18n = i18next.t('desk_ca_name_missing');
            errors.push(ca.resourceName ? i18n + ' - ' + ca.resourceName : i18n);
        }else{
            var resourceName = ca.resourceName || '';
            var customActionName = ca.customActionName || '';
            if(customActionName.length > 250){
                errors.push(i18next.t('desk_ca_name_maxlength') + ' - ' + (resourceName || customActionName));
            }    
        }
        if(!ca.hasOwnProperty('targetId')){
            var i18n = i18next.t('desk_ca_targetId_missing');
            errors.push(ca.resourceName ? i18n + ' - ' + ca.resourceName : i18n);
        }
        /** Custom Action Input field validation */
        var inputFields = ca.inputFields || [];
        let validFieldTypes = [
            "STRING","LOOKUP","TEXT","NUMBER","DECIMAL","PICKLIST","URL",
            "MULTIPICKLIST","CONTENT","BOOLEAN","DYNAMICPICKLIST","ARRAY","OBJECT"
        ]
        inputFields.forEach(inputField=>{
            /**display name valdiation*/
            if(!inputField.hasOwnProperty('displayName')){
                var i18n = i18next.t('desk_ca_inpfield_disname_missing');
                errors.push(ca.resourceName ? i18n + ' - ' + ca.resourceName : i18n);
            }else{
                if(inputField.displayName.length > 250){
                    errors.push(i18next.t('desk_ca_inpfield_disname_maxlength') + ' - ' + inputField.displayName);
                }
            }
            if(!inputField.hasOwnProperty('type')){
                var i18n = i18next.t('desk_ca_inpfield_type_missing');
                errors.push(i18n + ' - ' + inputField.displayName);
            } else if(validFieldTypes.indexOf(inputField.type) === -1){
                var i18n = i18next.t('desk_ca_inpfield_type_invalid');
                errors.push(i18n + ' - ' + inputField.displayName);
            }
        })
    })
    return errors;
}

function validateResourcesJSONTargets(targets){
    var errors = [];
    targets.forEach(target=>{
        if(target.hasOwnProperty('resourceName')){
            var resourceName = target.resourceName || '';
            if(resourceName.length > 250){
                errors.push(i18next.t('desk_ca_reasourcename_maxlength'));
            }
        }
        /** Custom Action Name validation */
        if(!target.hasOwnProperty('targetName')){
            var i18n = i18next.t('desk_target_name_missing');
            errors.push(target.resourceName ? i18n + ' - ' + target.resourceName : i18n);
        }else{
            var resourceName = target.resourceName || '';
            var targetName = target.targetName || '';
            if(targetName.length > 250){
                errors.push(i18next.t('desk_target_name_maxlength') + ' - ' + (resourceName || targetName));
            }
        }
        if(!target.hasOwnProperty('url')){
            var i18n = i18next.t('desk_target_url_missing');
            errors.push(target.resourceName ? i18n + ' - ' + target.resourceName : i18n);
        }if(!target.hasOwnProperty('httpMethod')){
            var i18n = i18next.t('desk_target_httpmethod_missing');
            errors.push(target.resourceName ? i18n + ' - ' + target.resourceName : i18n);
        }else{
            var validMethods = ["GET","POST","PUT","PATCH","DELETE"];
            if(validMethods.indexOf(target.httpMethod) === -1){
                var i18n = i18next.t('desk_target_httpmethod_invalid');
                errors.push(target.resourceName ? i18n + ' - ' + target.resourceName : i18n);
            }
        }
    })
    return errors;
}

function validateResourcesJson(resourcejson, projectRootDir) {
    var errors = [];
    var channel = resourcejson.channel;
    var permissions=resourcejson.permissions;
    var webhook=resourcejson.webhook;
    var fields=resourcejson.fields;
    var customActions=resourcejson.customActions;
    var targets=resourcejson.targets;
    if (!isUndefined(channel)) {
        var logopath = channel.channelLogoPath;
        if (isUndefined(logopath)) {
            errors.push(i18next.t("desk_resource_missinglogopath"));
        } else if (!isString(logopath)) {
            errors.push(i18next.t('desk_resource_invalidlogopath'));
        } else {
            let extnName = path.extname(logopath);
            let serviceName = service.getServiceName(projectRootDir);
            if (CONST.ALLOWED_EXTNS[serviceName].images.indexOf(extnName) === -1) {
                errors.push(i18next.t('desk_resource_invalidlogopathextn'));
            }
            if (logopath.indexOf('/') !== 0) {
                errors.push(i18next.t('desk_resource_invalidlogopathurl'));
            }
            if (!fs.existsSync(path.join(projectRootDir, logopath))) {
                errors.push(i18next.t('desk_resource_missinglogopathfile'));
            }
        }

        var contentTypes = channel.contentTypes;
        var contenttypearr = ["text/html", "text/plain"];
        if(!isUndefined(contentTypes)) {
            if(!Array.isArray(contentTypes)) {
                errors.push(i18next.t("desk_resource_invalidcontentType"));
            } else {
                contentTypes.forEach((element,i) => {
                    if(contenttypearr.indexOf(element) === -1) {
                        errors.push(i18next.t("desk_resource_invalidcontentTypelist")+ " at posistion "+ i);
                    }
                });
            } 
            
        }

        var redirectUrl = channel.redirectUrl;
        if(!isUndefined(redirectUrl) && !isValidURL(redirectUrl)) {
            errors.push(i18next.t("desk_resource_invalidredirecturl"));
        }

        var sync = channel.sync;
        if(!isUndefined(sync)) {
            if(!isObject(sync)){
                errors.push(i18next.t("desk_resource_invalidsync"));
             } 
            //else {
            //     if(!isUndefined(sync.push) && !isValidURL(sync.push)) {
            //         errors.push(i18next.t("desk_resource_invalidsynctype"));
            //     }
            //     if(!isUndefined(sync.pull) && !isValidURL(sync.pull)) {
            //         errors.push(i18next.t("desk_resource_invalidsynctype"));
            //     }
            // }
        }
    }

    if(!isUndefined(permissions)){
        if(!isArray(permissions)){
            errors.push(i18next.t("desk_resource_permission_invalid"));
        }else{
            if(permissions.length==0){
                errors.push(i18next.t("desk_resource_permission_empty"));
            }else if(permissions.length<=10){
                errors = errors.concat(validateResourcesJsonPermissions(permissions));
            }else{
                errors.push(i18next.t("desk_resource_permission_maxcount"));
            }
        }
    }

    if(!isUndefined(fields)){
        if(isArray(fields)){
            if(fields.length==0){
                errors.push(i18next.t("desk_resource_fields_empty"));
            }else if(fields.length > 10){
                errors.push(i18next.t("desk_resource_fields_maxcount"));
            }
        }else{
            errors.push(i18next.t("desk_resource_fields_invalid"));
        }
    } 

    let resourceNameValidationErrors = validateResourceJSONNames({customActions, fields, targets, webhook});
    if(resourceNameValidationErrors.length){
        errors = errors.concat(resourceNameValidationErrors);
    }

    if(!isUndefined(webhook)){
        if(isObject(webhook)){
            errors = errors.concat(validateResourcesJsonWebhook(webhook));
        }else{
            errors.push(i18next.t("desk_resource_webhook_empty"));
        }
    }   

    if(!isUndefined(customActions)){
        if(isArray(customActions)){
            errors = errors.concat(validateResourcesJSONCA(customActions));
        }else{
            errors.push(i18next.t("desk_resource_customactions_invalid"));
        }
    }
    if(!isUndefined(targets)){
        if(isArray(targets)){
            errors = errors.concat(validateResourcesJSONTargets(targets));
        }else{
            errors.push(i18next.t("desk_resource_targets_invalid"));
        }	
    } 

    return errors;
}

function validateTPAOAuthConfig(configs) {
    var errors = [], nameArray = [];

    for (var i = 0; i < configs.length; i++) {
        var config = configs[i];
        if (!isObject(config)) {
            errors.push(i18next.t('desk_manifest_tpo_config_invalidtype'));
            continue;
        }
        if( isObject(config) && Object.keys(config).length === 0 ) {
            errors.push(i18next.t('desk_manifest_tpo_config_empty'));
            continue;
        }

        var name = config.name;
        var clientID = config.clientId;
        var clientSecret = config.clientSecret;
        var authoriseURL = config.authoriseUrl;
        var tokenURL = config.tokenUrl;
        var revokeURL = config.revokeUrl;
        var params = config.params;

        if (isUndefined(name)) {
            errors.push(i18next.t('desk_manifest_tpo_name_missing'));
        } else {
            // unique 'name' validation
            if( !isString(name) || name === '' ) {
                errors.push(i18next.t('desk_manifest_tpo_name_empty'));
            } else if (!isAllowedString(name)) {
                errors.push(i18next.t('desk_manifest_tpo_name_invalidtype'));
            } else if (nameArray.indexOf(name.trim()) >= 0) {
                errors.push(i18next.t("desk_manifest_tpo_name_alreadyexists"));
            } else {
                nameArray.push(name.trim());
            }
        }


        if( isUndefined(clientID) ) {
            errors.push(i18next.t('desk_manifest_tpo_clientid_missing'));
        } else if( clientID === "" ) {
            errors.push(i18next.t('desk_manifest_tpo_clientid_invalidtype'));
        }

        if( isUndefined(clientSecret) ) {
            errors.push(i18next.t('desk_manifest_tpo_clientsecret_missing'));
        } else if( clientSecret === "" ) {
            errors.push(i18next.t('desk_manifest_tpo_clientsecret_invalidtype'));
        }

        if( isUndefined(authoriseURL) ) {
            errors.push(i18next.t('desk_manifest_tpo_authoriseurl_missing'));
        } else if( authoriseURL === "" || !isValidURL(authoriseURL)) {
            errors.push(i18next.t('desk_manifest_tpo_authoriseurl_invalidurl'));
        }

        if( isUndefined(tokenURL) ) {
            errors.push(i18next.t('desk_manifest_tpo_tokenurl_missing'));
        } else if( tokenURL === "" || !isValidURL(tokenURL)) {
            errors.push(i18next.t('desk_manifest_tpo_tokenurl_invalidurl'));
        }

        if( !isUndefined(revokeURL) && !isValidURL(revokeURL)) {
            errors.push(i18next.t('desk_manifest_tpo_revokeurl_invalidurl'));
        }

        if( !isUndefined(params) && !isObject(params) ) {
            errors.push(i18next.t('desk_manifest_tpo_params_invalidtype'));
        }
    }

    return errors;
}
function validateDeskOAuthConfig(config) {
    var errors = [];

    var clientID = config.clientId;
    var name = config.name;
    var clientSecret = config.clientSecret;
    var scopes = config.scopes;
    var params = config.params;

    if( !clientID ) {
        errors.push(i18next.t('desk_manifest_doc_clientid_missing'));
    } else if( !isString(clientID) )  {
        errors.push(i18next.t('desk_manifest_doc_clientid_invalidtype'));
    }

    if( !name ) {
        errors.push(i18next.t('desk_manifest_doc_name_missing'));
    } else if( !isString(name)) {
        errors.push(i18next.t('desk_manifest_doc_name_invalidtype'));
    }

    if( !clientSecret ) {
        errors.push(i18next.t('desk_manifest_doc_clientsecret_missing'));
    } else if( !isString(clientSecret) ) {
        errors.push(i18next.t('desk_manifest_doc_clientsecret_invalidtype'));
    }

    if(scopes && !Array.isArray(scopes)) {
        errors.push(i18next.t('desk_manifest_doc_scopes_invalidtype'));
    }
    if(params && !isObject(params)) {
        errors.push(i18next.t('desk_manifest_doc_params_invalidtype'));
    }

    return errors;
}
function validateConfigOptions(configs) {
    var errors = [], nameArray = [];

    for (var i = 0; i < configs.length; i++) {
        var config = configs[i];
        if (!isObject(config)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidconfigtype'));
            continue;
        }
        if( isObject(config) && Object.keys(config).length === 0 ) {
            errors.push(i18next.t('desk_pluginmanifest_emptyconfig'));
            continue;
        }

        var name = config.name;
        var defaultvalue = config.default;
        var type = config.type;
        var userdefined = config.userdefined;
        var mandatory = config.mandatory;
        var secure = config.secure;
        var displayName = config.displayName;
        var description = config.description;

        if (isUndefined(name)) {
            errors.push(i18next.t('desk_pluginmanifest_missingconfigname'));
        } else {
            if (!isString(name) || ( isString(name) && name.trim() === "" ) ) {
                errors.push(i18next.t('desk_pluginmanifest_invalidconfigname'));
            }
            // unique 'name' validation
            if (!/^[a-zA-Z][a-zA-Z0-9-_ ]*$/i.test(name)) {
                errors.push(i18next.t("desk_pluginmanifest_invalidcharacters") + "'config' : " + name);
            } else if (isString(name) && nameArray.indexOf(name.trim()) >= 0) {
                errors.push(i18next.t("desk_pluginmanifest_duplicatename") + "'config' : " + name);
            } else {
                nameArray.push(name.trim());
            }
        }

        if(!isUndefined(displayName)) {
            if(!isString(displayName)) {
                errors.push(i18next.t("desk_pluginmanifest_invaliddisplayname"));
            } else if(displayName.length > 255){
                errors.push(i18next.t("desk_pluginmanifest_invaliddisplaynamelength"));
            }
        }
        if(!isUndefined(description)) {
            if(!isString(description)) {
                errors.push(i18next.t("desk_pluginmanifest_invaliddescription"));
            } else if(description.length > 255){
                errors.push(i18next.t("desk_pluginmanifest_invaliddescriptionlength"));
            }
        }
        // var undefinedList = [type, userdefined, mandatory, secure].filter(isUndefined);
        var configType = ["text", "number", "radio", "switch", "picklist", "multiselect", "checkbox"];
        // if (undefinedList.length > 0 && undefinedList.length < 4) {
        //     errors.push(i18next.t('desk_pluginmanifest_requiredconfigproperties') + name);
        //     continue;
        // }
        if(!isUndefined(userdefined)){
            if (!isBoolean(userdefined)) {
                errors.push(i18next.t('desk_pluginmanifest_invaliduserdefinedconfigtype'));
            }
            if (userdefined) {
                if(!isUndefined(type)){
                    if (!isString(type)) {
                        errors.push(i18next.t('projects_pluginmanifest_invalidtypeinconfig'));
                        //onfigType.indexOf(type) === -1
                    } else if (configType.indexOf(type) === -1){
                        errors.push(i18next.t('desk_pluginmanifest_invalidlistoftypeinconfig'));
                    }
                }else{
                    errors.push(i18next.t('desk_pluginmanifest_requiredconfigproperties_type') + name);
                }
                
                if(!isUndefined(mandatory)){
                    if (!isBoolean(mandatory)) {
                        errors.push(i18next.t('desk_pluginmanifest_invalidmandatorytype'));
                    }
                }else{
                    errors.push(i18next.t('desk_pluginmanifest_requiredconfigproperties_mandatory') + name);
                }
                if(!isUndefined(secure)){
                    if (!isBoolean(secure)) {
                        errors.push(i18next.t('desk_pluginmanifest_invalidsecuretype'));
                    }
                }else{
                    errors.push(i18next.t('desk_pluginmanifest_requiredconfigproperties_secure') + name);
                }

            }
        }   
    }
    return errors;
}

function validateConnectorConfig(connectors) {
    var errors = [], nameArray = [], serviceArray = [];

    for (var i = 0; i < connectors.length; i++) {
        var connector = connectors[i];

        if (!isObject(connector)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidconnectorstype'));
            continue;
        }
        if (isObject(connector) && Object.keys(connector).length === 0) {
            errors.push(i18next.t('desk_manifest_connectors_emptyobject'));
            continue;
        }

        var name = connector.connectionName;
        var service = connector.serviceName;
        var connectionLink = connector.connectionLinkName;
        var scope = connector.scope;

        if (isUndefined(name)) {
            errors.push(i18next.t('desk_pluginmanifest_missingconnectorname'));
        } else if (!isString(name) || ( isString(name) && name.trim() === "") )  {
            errors.push(i18next.t('desk_pluginmanifest_invalidconnectorname'));
        } else {

            // unique 'name' validation
            if (!/^[a-zA-Z][a-zA-Z0-9-_ ]*$/i.test(name)) {
                errors.push(i18next.t("desk_pluginmanifest_invalidcharacters") + "'connectors' : " + name);
            } else if (nameArray.indexOf(name.trim()) >= 0) {
                errors.push(i18next.t("desk_pluginmanifest_duplicatename") + "'connectors' : " + name);
            }
            nameArray.push(name.trim());

            // Servicename property check
            if (isUndefined(service)) {
                errors.push(i18next.t('desk_pluginmanifest_missingconnectorservice') + name);
            } else {
                if (!isString(service) || ( isString(service) && service.trim() === '') ) {
                    errors.push(i18next.t('desk_pluginmanifest_invalidconnectorservice') + name);
                } else if(service === "zlabs_integration") {
                    errors.push(i18next.t('desk_pluginmanifest_zlabsservicename') + name);
                } else if (serviceArray.indexOf(service.trim()) >= 0) {
                    errors.push(i18next.t("desk_pluginmanifest_duplicateservicename") + "'connectors' : " + name);
                }
                serviceArray.push(service.trim());
            }
            
            // ConnectionLinkname property check
            if (isUndefined(connectionLink)) {
                errors.push(i18next.t('desk_manifest_connectors_connectionLink_missing'));
            } else if (!isString(connectionLink)) {
                errors.push(i18next.t('desk_manifest_connectors_connectionLink_invalid'));
            }

            if (!isUndefined(scope)) {
                if (isArray(scope)) {
                    for (var j = 0; j < scope.length; j++) {
                        if (!isString(scope[j])) {
                            errors.push(i18next.t('desk_pluginmanifest_invalidscopearraytype') + name);
                            break;
                        }
                        if(scope[j].trim()===''){
                            errors.push(i18next.t('desk_pluginmanifest_scopeisempty'));
                        }
                    }
                } else {
                    errors.push(i18next.t('desk_pluginmanifest_invalidconnectorscope') + name);
                }
            }
        }
    }
    return errors;
}

function validateDomains(domainsList) {
    var invalidURLs = domainsList.filter(function (domain) {
        return !util.isValidURL(domain);
    });

    return invalidURLs && invalidURLs.length > 0;
}

function validateWidgetConfigs(projectRootDir, widgets) {
    var errs = [], nameArray = [], oneLocationArray = [ "desk.background", "desk.extension.preference", "desk.extension.telephony" ];

    for (var i = 0; i < widgets.length; i++) {
        var widget = widgets[i];
        var name = widget.name;
        if (isUndefined(name)) {
            errs.push(i18next.t('desk_pluginmanifest_missingwidgetname') + name);
        } else {
            if ( !isString(name) || ( isString(name) && name.trim() === "" ) ) {
                errs.push(i18next.t('desk_pluginmanifest_invalidwidgetname') + name);
            }
            if (!/^[a-zA-Z][a-zA-Z0-9-_ ]*$/i.test(name)) {
                errs.push(i18next.t("desk_pluginmanifest_invalidcharacters") + "'widgets' : " + name);
            } else if (isString(name) && nameArray.indexOf(name.trim()) >= 0) {
                errs.push(i18next.t("desk_pluginmanifest_duplicatename") + "'widgets' : " + name);
            } else if( isString(name) ) {
                nameArray.push(name.trim());
            }
        }

        var url = widget.url;
        var logo = widget.logo;
        var icon = widget.icon;

        if (isUndefined(url)) {
            errs.push(i18next.t('desk_pluginmanifest_missingurlinwidget') + name);
        } else if (!util.isValidURL(url)) {
            if (isString(url)) {
                var startsWithSlash = url.indexOf('/') === 0;
                if (startsWithSlash) {
                    if (!fs.existsSync(path.join(projectRootDir, url))) {
                        errs.push(i18next.t('desk_pluginmanifest_missingfileinappsource'));
                    }
                } else if(url.indexOf('{') != 0) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidurl') + name);
                }
            } else {
                errs.push(i18next.t('desk_pluginmanifest_invalidurltype') + name);
            }
        }

        let serviceName = service.getServiceName(projectRootDir);

        if (!isUndefined(logo)) {
            if (isString(logo)) {
                let extnName = path.extname(logo);
                if (CONST.ALLOWED_EXTNS[serviceName].images.indexOf(extnName) === -1) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidlogoextn'));
                }
                if (logo.indexOf('/') !== 0) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidlogo'));
                }
                if (!fs.existsSync(path.join(projectRootDir, logo))) {
                    errs.push(i18next.t('desk_pluginmanifest_missinglogofile'));
                }
            } else {
                errs.push(i18next.t('deskpluginmanifest_invalidlogotype'));
            }
        }
        
        if (!isUndefined(icon)) {
            if (isString(icon)) {
                let extnName = path.extname(icon);
                if (CONST.ALLOWED_EXTNS[serviceName].images.indexOf(extnName) === -1) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidiconextn'));
                }
                if (icon.indexOf('/') !== 0) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidicon'));
                }
                if (!fs.existsSync(path.join(projectRootDir, icon))) {
                    errs.push(i18next.t('desk_pluginmanifest_missingiconfile'));
                }
            } else {
                errs.push(i18next.t('deskpluginmanifest_invalidicontype'));
            }
        }
    }
    return errs;
}

function validateResourcesJsonPermissions(permissions){
    var nameArray=[];var errs = [];
    for(var i=0;i<permissions.length;i++){
        var permission=permissions[i];
        var name=permission.apiName;
        var displayName=permission.displayName;
        var description=permission.description;
        var defaultState=permission.defaultState;
        if (isUndefined(name)) {
            errs.push(i18next.t('desk_resource_missingapiname') );
        } else {
            if ( !isString(name) || ( isString(name) && name.trim() === "" ) ) {
                errs.push(i18next.t('desk_resource_invalidapiname') );
            }
            if (isString(name) && nameArray.indexOf(name.trim()) >= 0) {
                errs.push(i18next.t("desk_resource_duplicatename") + "'resources.json' : " + name);
            } else if( isString(name) ) {
                nameArray.push(name.trim());
            }
        }

        if (isUndefined(displayName)) {
            errs.push(i18next.t('desk_resource_missingdisplayname') + name);
        } else if ( !isString(displayName) || ( isString(displayName) && displayName.trim() === "" ) ) {
                errs.push(i18next.t('desk_resource_invaliddisplayname') + name);
        }
         
        if (isUndefined(description)) {
            errs.push(i18next.t('desk_resource_missingdescription') + name);
        } else if ( !isString(description) || ( isString(description) && description.trim() === "" ) ) {
            errs.push(i18next.t('desk_resource_invaliddescription') + name);
        }

        if (isUndefined(defaultState)) {
            errs.push(i18next.t('desk_resource_missing_defaultState') + name);
        } else if (!isBoolean(defaultState)) {
                errs.push(i18next.t('desk_resource_invalid_defaultState') + name);
        }
    }
    return errs;
}

function validateResourcesJsonWebhook(webhook){
    var errs=[];
    var resourceName = webhook.resourceName;
    var url=webhook.url;
    var name=webhook.name;
    var subscriptions=webhook.subscriptions;
    var ignoreSourceId=webhook.ignoreSourceId;
    if(isUndefined(url)){
        errs.push(i18next.t('desk_resource_missing_webhookurl') );
    }else if ( !isString(url) || ( isString(url) && url.trim() === "" ) ) {
        errs.push(i18next.t('desk_resource_invalid_webhookurl') );
    }

    if(isUndefined(name)){
        errs.push(i18next.t('desk_resource_missing_webhookname') );
    }else if ( !isString(name) || ( isString(name) && name.trim() === "" ) ) {
        errs.push(i18next.t('desk_resource_invalid_webhookname') );
    }

    if(isUndefined(subscriptions)){
        errs.push(i18next.t('desk_resource_missing_webhooksubscriptions') );
    }else if(!isObject(subscriptions)){
        errs.push(i18next.t('desk_resource_invalid_webhooksubscriptions') );
    }

    if(isUndefined(ignoreSourceId)){
        errs.push(i18next.t('desk_resource_missing_webhookignoreSourceId') );
    }else if ( !isString(ignoreSourceId) ) {
        errs.push(i18next.t('desk_resource_invalid_webhookignoreSourceId') );
    }

    return errs;
}

function validateResourceJSONNames(resourceJson){
    var errs = [];
    Object.keys(resourceJson).reduce((result, resourceKey)=>{
        let resourceData = resourceJson[resourceKey];
        let {resourceNames:nameList} = result;

        var validateNames = function(source){
            if(isUndefined(source.resourceName)){
                errs.push(i18next.t('desk_resource_missing_'+resourceKey.toLowerCase()+'ResourceName'));
            }else if ( !isString(source.resourceName) || ( isString(source.resourceName) && source.resourceName.trim() === "" ) ) {
                errs.push(i18next.t('desk_resource_invalid_'+resourceKey.toLowerCase()+'ResourceName'));
            }else{
                if(nameList.indexOf(source.resourceName) !== -1){
                    errs.push(i18next.t("desk_resourcename_duplicatename") + "'resources.json' : " + source.resourceName);
                }
            }
            return errs;
        }

        if(Array.isArray(resourceData)){
            resourceData.forEach((resource)=>{
                if(isObject(resourceData)){
                    errs = validateNames(resource);
                    if(!errs.length){
                        result.resourceNames = [...result.resourceNames, resource.resourceName];
                    }
                }
            })
        }else if(isObject(resourceData)){
            errs = validateNames(resourceData);
            if(!errs.length){
                result.resourceNames = [...result.resourceNames, resourceData.resourceName];
            }
        }
        return result;
    },{resourceNames : []});
    return errs;
}

module.exports = {
    name: i18next.t('pluginmanifestrule_name'),
    run: function (projectRootDir, fileTree) {
        return validatePluginManifest(projectRootDir);
    }
};
