// zoho-extension-toolkit/lib/rules/MAIL.PluginManifestRule.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/

var fs = require('fs');
var path = require('path');
var i18next = require('i18next');
var CONST = require('../commands/constants.js');
var util = require('../utils');
var service = require('./../utils/service.js');
var isString = util.isString;
var isUndefined = util.isUndefined;
var isObject = util.isObject;
var isArray = util.isArray;
var isBoolean = util.isBoolean;

var MANIFEST_KEY = [ "service", "modules", "whiteListedDomains" ];
var LOCATION_CONSTANT = [ "zoho.mail.preview.rightpanel", "zoho.mail.attachment.popup" ];

var CSP_DOMAINS_ARR = ["connect-src", "img-src", "style-src", "script-src"];
let cspArrayLength = 10;

function validatePluginManifest(projectRootDir) {

    var errors = [];
    var pluginManifest = require(path.join(projectRootDir, CONST.PLUGIN_MANIFEST));

    let manifestKeys = Object.keys(pluginManifest);

    //plugin manifest mandatory key validation
    for (var i = 0; i < MANIFEST_KEY.length; i++) {
        if (manifestKeys.indexOf(MANIFEST_KEY[i]) === -1) {
            errors.push("\"" + MANIFEST_KEY[i] + "\"" + i18next.t('pluginmanifest_keymissing'));
        }
    }

    //locale file exists validation
    var manifestLocale = pluginManifest.locale;
    if (manifestLocale) {
        if( isArray(manifestLocale) ) {
            var locales = manifestLocale;
            for( var i=0;i<locales.length; i++) {
                let _locale = locales[i];
                let relativePath = path.join( 'app', 'translations', _locale + '.json');
                let localePath = path.join(projectRootDir, relativePath);
                if (!fs.existsSync(localePath)) { 
                    errors.push(i18next.t('desk_pluginmanifest_localenotfound'));
                }
            }
        } else {
            errors.push(i18next.t('desk_pluginmanifest_locale_invalidtype'));
        }
        var defaultLocale = pluginManifest.defaultLocale;
        if(isUndefined(defaultLocale)) {
            errors.push(i18next.t('mail_pluginmanifest_missingdefaultlocale'));
        } else if(!isString(defaultLocale) || defaultLocale.trim() === "") {
            errors.push(i18next.t('mail_pluginmanifest_invaliddefaultlocale'));
        }
    }

    //storage type check
    var storage = pluginManifest.storage;
    if (!isUndefined(storage) && typeof storage !== 'boolean' ) {
        errors.push(i18next.t('desk_pluginmanifest_invalidstorage'));
    }

    //module validation
    var modules = pluginManifest.modules;
    if (!isUndefined(modules)) {
        if (!isObject(modules)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidmodules'));
        } else if (isUndefined(modules.widgets)) {
            errors.push(i18next.t('desk_pluginmanifest_widgetkeymissing'));
        } else if (!isArray(modules.widgets)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidwidgettype'));
        } else if (isArray(modules.widgets) && modules.widgets.length === 0) {
            errors.push(i18next.t('desk_pluginmanifest_widgetisempty'));
        } else {
            errors = errors.concat(validateWidgetConfigs(projectRootDir, modules.widgets));
        }
    }

    var whiteListedDomains = pluginManifest.whiteListedDomains;
    if (whiteListedDomains) {
        if (!isArray(whiteListedDomains)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidwhitelisteddomaintype'));
        } else {
            if (validateDomains(whiteListedDomains)) {
                errors.push(i18next.t('desk_pluginmanifest_invalidwhitelisteddomain'));
            }
        }
    }

    // 'connectors' validation
    var connectorsConfig = pluginManifest.connectors;
    if (!isUndefined(connectorsConfig)) {
        if (isArray(connectorsConfig)) {
            errors = errors.concat(validateConnectorConfig(connectorsConfig));
        } else {
            errors.push(i18next.t('desk_pluginmanifest_invalidconnectortype'));
        }
    }

    // 'config' validations
    var configs = pluginManifest.config;
    if (!isUndefined(configs)) {
        if (isArray(configs)) {
            if (configs.length > 20) {
                errors.push(i18next.t('desk_pluginmanifest_configlimitexceeded'));
            } else {
                errors = errors.concat(validateConfigOptions(configs));
            }
        } else {
            errors.push(i18next.t('desk_pluginmanifest_invalidconfigtype'));
        }
    }
    // 'cspDomains' validations
    if (pluginManifest.cspDomains) {
        let cspDomains = pluginManifest.cspDomains;
        errors = errors.concat(service.validateCspDomains(cspDomains));
    }

    return errors;

}

function validateConfigOptions(configs) {
    var errors = [], nameArray = [];

    for (var i = 0; i < configs.length; i++) {
        var config = configs[i];
        if (!isObject(config)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidconfigtype'));
            continue;
        }
        if( isObject(config) && Object.keys(config).length === 0 ) {
            errors.push(i18next.t('desk_pluginmanifest_emptyconfig'));
            continue;
        }

        var name = config.name;
        var type = config.inputType;
        var mandatory = config.mandatory;
        var secure = config.secure;

        if (isUndefined(name)) {
            errors.push(i18next.t('desk_pluginmanifest_missingconfigname'));
        } else {
            if (!isString(name) || ( isString(name) && name.trim() === "" ) ) {
                errors.push(i18next.t('desk_pluginmanifest_invalidconfigname'));
            }
            // unique 'name' validation
            if (!/^[a-zA-Z][a-zA-Z0-9-_ ]*$/i.test(name)) {
                errors.push(i18next.t("desk_pluginmanifest_invalidcharacters") + "'config' : " + name);
            } else if (isString(name) && nameArray.indexOf(name.trim()) >= 0) {
                errors.push(i18next.t("desk_pluginmanifest_duplicatename") + "'config' : " + name);
            } else {
                nameArray.push(name.trim());
            }
        }

        var undefinedList = [type, mandatory, secure].filter(isUndefined);
        var configType = ["text", "dropdown", "radio"];

        if (undefinedList.length > 0) {
            errors.push(i18next.t('mail_pluginmanifest_requiredconfigproperties') + name);
            continue;
        } else {
            if (!isString(type)) {
                errors.push(i18next.t('projects_pluginmanifest_invalidtypeinconfig'));
            } else if (configType.indexOf(type) === -1){
                errors.push(i18next.t('mail_pluginmanifest_invalidlistoftypeinconfig'));
            }
            if (!isBoolean(mandatory)) {
                errors.push(i18next.t('mail_pluginmanifest_invalidmandatorytype'));
            }
            if (!isBoolean(secure)) {
                errors.push(i18next.t('mail_pluginmanifest_invalidsecuretype'));
            }
        }
    }
    return errors;
}

function validateDomains(domainsList) {
    var invalidURLs = domainsList.filter(function (domain) {
        return !util.isValidURL(domain);
    });

    return invalidURLs && invalidURLs.length > 0;
}

function validateConnectorConfig(connectors) {
    var errors = [];

    for (var i = 0; i < connectors.length; i++) {
        var connector = connectors[i];

        if (!isObject(connector)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidconnectorstype'));
            continue;
        }
        if (isObject(connector) && Object.keys(connector).length === 0) {
            errors.push(i18next.t('desk_manifest_connectors_emptyobject'));
            continue;
        }

        var service = connector.serviceName;

        // Servicename property check
        if (isUndefined(service)) {
            errors.push(i18next.t('desk_pluginmanifest_missingconnectorservice') + name);
        } else if (!isString(service) || ( isString(service) && service.trim() === '') ) {
            errors.push(i18next.t('desk_pluginmanifest_invalidconnectorservice') + name);
        }
    }
    return errors;
}

function validateWidgetConfigs(projectRootDir, widgets) {
    var errs = [], nameArray = [];

    for (var i = 0; i < widgets.length; i++) {
        var widget = widgets[i];
        var name = widget.name;
        if (isUndefined(name)) {
            errs.push(i18next.t('desk_pluginmanifest_missingwidgetname') + name);
        } else {
            if ( !isString(name) || ( isString(name) && name.trim() === "" ) ) {
                errs.push(i18next.t('desk_pluginmanifest_invalidwidgetname') + name);
            }
            if (!/^[a-zA-Z][a-zA-Z0-9-_ ]*$/i.test(name)) {
                errs.push(i18next.t("desk_pluginmanifest_invalidcharacters") + "'widgets' : " + name);
            } else if (isString(name) && nameArray.indexOf(name.trim()) >= 0) {
                errs.push(i18next.t("desk_pluginmanifest_duplicatename") + "'widgets' : " + name);
            } else if( isString(name) ) {
                nameArray.push(name.trim());
            }
        }

        var location = widget.location;
        var url = widget.url;
        var logo = widget.logo;
        var icon = widget.icon;

        if (isUndefined(location)) {
            errs.push(i18next.t('desk_pluginmanifest_missinglocationinwidget') + name);
        } else if (LOCATION_CONSTANT.indexOf(location) === -1) {
            errs.push(i18next.t('desk_pluginmanifest_invalidlocation') + name);
        }

        if (isUndefined(url)) {
            errs.push(i18next.t('desk_pluginmanifest_missingurlinwidget') + name);
        } else if (util.isValidURL(url)) {
            var parsedURL = new URL(url);
            if(parsedURL.hostname.indexOf(".zoho.") === -1 && parsedURL.hostname.indexOf(".localzoho.") === -1){
                errs.push(i18next.t('mail_pluginmanifest_isnotsubdomain') + name);
            }
        } else {
            if (isString(url)) {
                var startsWithSlash = url.indexOf('/') === 0;
                if (startsWithSlash) {
                    if (!fs.existsSync(path.join(projectRootDir, url))) {
                        errs.push(i18next.t('desk_pluginmanifest_missingfileinappsource'));
                    }
                } else {
                    errs.push(i18next.t('desk_pluginmanifest_invalidurl') + name);
                }
            } else {
                errs.push(i18next.t('desk_pluginmanifest_invalidurltype') + name);
            }
        }

        let serviceName = service.getServiceName(projectRootDir);

        if (!isUndefined(logo)) {
            if (isString(logo)) {
                let extnName = path.extname(logo);
                if (CONST.ALLOWED_EXTNS[serviceName].images.indexOf(extnName) === -1) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidlogoextn'));
                }
                if (logo.indexOf('/') !== 0) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidlogo'));
                }
                if (!fs.existsSync(path.join(projectRootDir, logo))) {
                    errs.push(i18next.t('desk_pluginmanifest_missinglogofile'));
                }
            } else {
                errs.push(i18next.t('deskpluginmanifest_invalidlogotype'));
            }
        }

        if (!isUndefined(icon)) {
            if (isString(icon)) {
                let extnName = path.extname(icon);
                if (CONST.ALLOWED_EXTNS[serviceName].images.indexOf(extnName) === -1) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidiconextn'));
                }
                if (icon.indexOf('/') !== 0) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidicon'));
                }
                if (!fs.existsSync(path.join(projectRootDir, icon))) {
                    errs.push(i18next.t('desk_pluginmanifest_missingiconfile'));
                }
            } else {
                errs.push(i18next.t('deskpluginmanifest_invalidicontype'));
            }
        }
    }
    return errs;
}

module.exports = {
    name: i18next.t('pluginmanifestrule_name'),
    run: function (projectRootDir) {
        return validatePluginManifest(projectRootDir);
    }
}
