// zoho-extension-toolkit/lib/rules/ORCHESTLY.PluginManifestRule.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/

var path = require('path');
var fs = require('fs');
var i18next = require('i18next');
var util = require('../utils');
var CONST = require('../commands/constants');
var service = require('../utils/service');
var isUndefined = util.isUndefined;
var isBoolean = util.isBoolean;
var isObject = util.isObject;
var isString = util.isString;
var isArray = util.isArray;
var isAllowedString = util.isAllowedString;
var isValidURL = util.isValidURL;

var LOCATION_CONSTANT = [
    "job.detail.rightpanel", "job.detail.attachment.popup", 
    "job.detail.bottompanel", "job.detail.during.popup", "extension.settings", 
    "job.detail.comment.popup", "blueprint.detail.duringtab"
];
var MANIFEST_KEY = [
    "locale", "storage", "service", "modules", "connectors", "config", "whiteListedDomains"
];

var CSP_DOMAINS_ARR = ["connect-src", "img-src", "style-src", "script-src"];
let cspArrayLength = 10;

function validatePluginManifest(projectRootDir) {
    var errors = [];

    var manifest = require(path.join(projectRootDir, CONST.PLUGIN_MANIFEST));

    let manifestKeys = Object.keys(manifest);

    //plugin manifest mandatory key validation
    for (var i = 0; i < MANIFEST_KEY.length; i++) {
        if (manifestKeys.indexOf(MANIFEST_KEY[i]) === -1) {
            errors.push("\"" + MANIFEST_KEY[i] + "\"" + i18next.t('pluginmanifest_keymissing'));
        }
    }
    if( errors.length > 0 ) return errors;

    //locale file exists validation
    if (manifest.locale) {
        if( isArray(manifest.locale) ) {
            var locales = manifest.locale;
            for( var i=0;i<locales.length; i++) {
                let _locale = locales[i];
                let relativePath = path.join( 'app', 'translations', _locale + '.json');
                let localePath = path.join(projectRootDir, relativePath);
                if (!fs.existsSync(localePath)) {
                    errors.push(i18next.t('desk_pluginmanifest_localenotfound') + relativePath);
                }
            }
        } else {
            errors.push(i18next.t('desk_pluginmanifest_locale_invalidtype'));
        }
    }

    if (typeof manifest.storage !== 'boolean' ) {
        errors.push(i18next.t('desk_pluginmanifest_invalidstorage'));
    }

    if (!isUndefined(manifest.modules)) {
        if (!isObject(manifest.modules)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidmodules'));
        } else if (isUndefined(manifest.modules.widgets)) {
            errors.push(i18next.t('desk_pluginmanifest_widgetkeymissing'));
        } else if (!isArray(manifest.modules.widgets)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidwidgettype'));
        } else if (isArray(manifest.modules.widgets) && manifest.modules.widgets.length === 0) {
            errors.push(i18next.t('desk_pluginmanifest_widgetisempty'));
        } else {
            errors = errors.concat(validateWidgetConfigs(projectRootDir, manifest.modules.widgets));
        }
    }

    if (manifest.whiteListedDomains) {
        if (!isArray(manifest.whiteListedDomains)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidwhitelisteddomaintype'));
        } else {
            if (validateDomains(manifest.whiteListedDomains)) {
                errors.push(i18next.t('desk_pluginmanifest_invalidwhitelisteddomain'));
            }
        }
    }

    // 'connectors' validation
    var connectorsConfig = manifest.connectors;
    if (!isUndefined(connectorsConfig)) {
        if (isArray(connectorsConfig)) {
            errors = errors.concat(validateConnectorConfig(connectorsConfig));
        } else {
            errors.push(i18next.t('desk_pluginmanifest_invalidconnectortype'));
        }
    }

    // 'config' validations
    var configs = manifest.config;
    if (!isUndefined(configs)) {
        if (isArray(configs)) {
            if (configs.length > 20) {
                errors.push(i18next.t('desk_pluginmanifest_configlimitexceeded'));
            } else {
                errors = errors.concat(validateConfigOptions(configs));
            }
        } else {
            errors.push(i18next.t('desk_pluginmanifest_invalidconfigtype'));
        }
    }

    // 'cspDomains' validations
    if (manifest.cspDomains) {
        let cspDomains = manifest.cspDomains;
        if (!isObject(cspDomains) || cspDomains.constructor !== Object) {
            errors.push(i18next.t('pluginmanifest_invalidCSPDomains'));
        } else {             
            CSP_DOMAINS_ARR.every((element,i) => {
                if(element) {                     
                    if(cspDomains[element].length > cspArrayLength) {
                        errors.push(element + i18next.t('pluginmanifest_csp_size_exceeded') + cspArrayLength);
                    }      
                    if(isArray(cspDomains[element])) {
                        if(validateDomains(cspDomains[element])) {
                            errors.push(cspDomains[element] + i18next.t('pluginmanifest_invalidConnectElementEntry') + element);
                        }
                    }
                    else {
                        errors.push(element + i18next.t('pluginmanifest_invalidConnectArray'));
                        return false;
                    }
                }
            });                   
        }
    }

    return errors;
}

function validateConfigOptions(configs) {
    var errors = [], nameArray = [];

    for (var i = 0; i < configs.length; i++) {
        var config = configs[i];
        if (!isObject(config)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidconfigtype'));
            continue;
        }
        if( isObject(config) && Object.keys(config).length === 0 ) {
            errors.push(i18next.t('desk_pluginmanifest_emptyconfig'));
            continue;
        }

        var name = config.name;
        var defaultvalue = config.default;
        var type = config.type;
        var userdefined = config.userdefined;
        var mandatory = config.mandatory;
        var secure = config.secure;

        if (isUndefined(name)) {
            errors.push(i18next.t('desk_pluginmanifest_missingconfigname'));
        } else {
            if (!isString(name) || ( isString(name) && name.trim() === "" ) ) {
                errors.push(i18next.t('desk_pluginmanifest_invalidconfigname'));
            }
            // unique 'name' validation
            if (!/^[a-zA-Z][a-zA-Z0-9-_ ]*$/i.test(name)) {
                errors.push(i18next.t("desk_pluginmanifest_invalidcharacters") + "'config' : " + name);
            } else if (isString(name) && nameArray.indexOf(name.trim()) >= 0) {
                errors.push(i18next.t("desk_pluginmanifest_duplicatename") + "'config' : " + name);
            } else {
                nameArray.push(name.trim());
            }
        }

        var undefinedList = [type, userdefined, mandatory, secure].filter(isUndefined);
        var configType = ["text","customfield","email","password","checkbox","radio","number","multiline","select","checklist"];

        if (undefinedList.length > 0 && undefinedList.length < 4) {
            errors.push(i18next.t('desk_pluginmanifest_requiredconfigproperties') + name);
            continue;
        }
        if (undefinedList.length === 0) {
            if (!isString(type)) {
                errors.push(i18next.t('projects_pluginmanifest_invalidtypeinconfig'));
            } else if (configType.indexOf(type) === -1){
                errors.push(i18next.t('projects_pluginmanifest_invalidlistoftypeinconfig'));
            }
            if (!isBoolean(mandatory)) {
                errors.push(i18next.t('projects_pluginmanifest_invalidmandatorytype'));
            }
            if (!isBoolean(secure)) {
                errors.push(i18next.t('projects_pluginmanifest_invalidsecuretype'));
            }
        }
    }
    return errors;
}

function validateConnectorConfig(connectors) {
    var errors = [], nameArray = [];

    for (var i = 0; i < connectors.length; i++) {
        var connector = connectors[i];

        if (!isObject(connector)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidconnectorstype'));
            continue;
        }
        if (isObject(connector) && Object.keys(connector).length === 0) {
            errors.push(i18next.t('desk_manifest_connectors_emptyobject'));
            continue;
        }

        var name = connector.connectionName;
        var service = connector.serviceName;
        var connectionLink = connector.connectionLinkName;
        var scope = connector.scope;

        if (isUndefined(name)) {
            errors.push(i18next.t('desk_pluginmanifest_missingconnectorname'));
        } else if (!isString(name) || ( isString(name) && name.trim() === "") )  {
            errors.push(i18next.t('desk_pluginmanifest_invalidconnectorname'));
        } else {

            // unique 'name' validation
            if (!/^[a-zA-Z][a-zA-Z0-9-_ ]*$/i.test(name)) {
                errors.push(i18next.t("desk_pluginmanifest_invalidcharacters") + "'connectors' : " + name);
            } else if (nameArray.indexOf(name.trim()) >= 0) {
                errors.push(i18next.t("desk_pluginmanifest_duplicatename") + "'connectors' : " + name);
            }
            nameArray.push(name.trim());

            // Servicename property check
            if (isUndefined(service)) {
                errors.push(i18next.t('desk_pluginmanifest_missingconnectorservice') + name);
            } else if (!isString(service) || ( isString(service) && service.trim() === '') ) {
                errors.push(i18next.t('desk_pluginmanifest_invalidconnectorservice') + name);
            }

            // ConnectionLinkname property check
            if (isUndefined(connectionLink)) {
                errors.push(i18next.t('desk_manifest_connectors_connectionLink_missing'));
            } else if (!isString(connectionLink)) {
                errors.push(i18next.t('desk_manifest_connectors_connectionLink_invalid'));
            }

            if (!isUndefined(scope)) {
                if (isArray(scope)) {
                    for (var j = 0; j < scope.length; j++) {
                        if (!isString(scope[j])) {
                            errors.push(i18next.t('desk_pluginmanifest_invalidscopearraytype') + name);
                            break;
                        }
                        if(scope[j].trim()===''){
                            errors.push(i18next.t('desk_pluginmanifest_scopeisempty'));
                        }
                    }
                } else {
                    errors.push(i18next.t('desk_pluginmanifest_invalidconnectorscope') + name);
                }
            }
        }
    }
    return errors;
}

function validateDomains(domainsList) {
    var invalidURLs = domainsList.filter(function (domain) {
        return !util.isValidURL(domain);
    });

    return invalidURLs && invalidURLs.length > 0;
}

function validateWidgetConfigs(projectRootDir, widgets) {
    var errs = [], nameArray = [];

    for (var i = 0; i < widgets.length; i++) {
        var widget = widgets[i];
        var name = widget.name;
        if (isUndefined(name)) {
            errs.push(i18next.t('desk_pluginmanifest_missingwidgetname') + name);
        } else {
            if ( !isString(name) || ( isString(name) && name.trim() === "" ) ) {
                errs.push(i18next.t('desk_pluginmanifest_invalidwidgetname') + name);
            }
            if (name.length > 200) {
                errs.push(i18next.t("orchestly_pluginmanifest_widget_name_size_exceeded")+ " 'widgets' : " + name);    
            }    
            if (!/^[a-zA-Z][a-zA-Z0-9-_ ]*$/i.test(name)) {
                errs.push(i18next.t("desk_pluginmanifest_invalidcharacters") + "'widgets' : " + name);    
            } else if (isString(name) && nameArray.indexOf(name.trim()) >= 0) {
                errs.push(i18next.t("desk_pluginmanifest_duplicatename") + "'widgets' : " + name);
            } else if( isString(name) ) {
                nameArray.push(name.trim());
            }
        }

        var location = widget.location;
        var url = widget.url;
        var logo = widget.logo;
        var icon = widget.icon;

        if (isUndefined(location)) {
            errs.push(i18next.t('desk_pluginmanifest_missinglocationinwidget') + name);
        } else if (LOCATION_CONSTANT.indexOf(location) === -1) {
            errs.push(i18next.t('desk_pluginmanifest_invalidlocation') + name);
        }

        if (isUndefined(url)) {
            errs.push(i18next.t('desk_pluginmanifest_missingurlinwidget') + name);
        } else if (!util.isValidURL(url)) {
            if (isString(url)) {
                var startsWithSlash = url.indexOf('/') === 0;
                if (startsWithSlash) {
                    if (!fs.existsSync(path.join(projectRootDir, url))) {
                        errs.push(i18next.t('desk_pluginmanifest_missingfileinappsource'));
                    }
                } else {
                    errs.push(i18next.t('desk_pluginmanifest_invalidurl') + name);
                }
            } else {
                errs.push(i18next.t('desk_pluginmanifest_invalidurltype') + name);
            }
        }

        let serviceName = service.getServiceName(projectRootDir);

        if (!isUndefined(logo)) {
            if (isString(logo)) {
                let extnName = path.extname(logo);
                if (CONST.ALLOWED_EXTNS[serviceName].images.indexOf(extnName) === -1) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidlogoextn'));
                }
                if (logo.indexOf('/') !== 0) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidlogo'));
                }
                if (!fs.existsSync(path.join(projectRootDir, logo))) {
                    errs.push(i18next.t('desk_pluginmanifest_missinglogofile'));
                }
            } else {
                errs.push(i18next.t('deskpluginmanifest_invalidlogotype'));
            }
        }

        if (!isUndefined(icon)) {
            if (isString(icon)) {
                let extnName = path.extname(icon);
                if (CONST.ALLOWED_EXTNS[serviceName].images.indexOf(extnName) === -1) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidiconextn'));
                }
                if (icon.indexOf('/') !== 0) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidicon'));
                }
                if (!fs.existsSync(path.join(projectRootDir, icon))) {
                    errs.push(i18next.t('desk_pluginmanifest_missingiconfile'));
                }
            } else {
                errs.push(i18next.t('deskpluginmanifest_invalidicontype'));
            }
        }
    }
    return errs;
}

module.exports = {
    name: i18next.t('pluginmanifestrule_name'),
    run: function (projectRootDir, fileTree) {
        return validatePluginManifest(projectRootDir);
    }
};
