// zoho-extension-toolkit/lib/rules/PluginManifestRule.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/

var CONST = require('../commands/constants');
var fse = require('fs-extra');
var path = require('path');
var serviceUtil = require('../utils/service.js');
var i18next = require('i18next');
var errMsgArr = [];
var util = require('../utils');
var manifestValidations = {

  validate: function (projectRootDir) {

    errMsgArr = [];
    var manifestFile = path.join(projectRootDir, CONST.PLUGIN_MANIFEST);
    var manifestJSON = fse.readJsonSync(manifestFile);

    if (!manifestJSON) {
      errMsgArr.push(i18next.t('pluginmanifestrule_invalidformat'));
      return;
    }

    if (!manifestJSON.service) {
      errMsgArr.push(i18next.t('pluginmanifestrule_missingservice'));
      return;
    }

    if (manifestJSON.service === '') {
      errMsgArr.push(i18next.t('pluginmanifestrule_emptyservice'));
      return;
    }
    var serviceNames = serviceUtil.getServiceNames().map(function (obj) { return obj.value; });
    if (serviceNames.indexOf(manifestJSON.service) === -1) {
      errMsgArr.push(i18next.t('pluginmanifestrule_invalidservice') + serviceNames.join(', '));
      return;
    }

    if (manifestJSON.service === CONST.SERVICE.CRM || manifestJSON.service === CONST.SERVICE.CREATOR) {

      if (!(manifestJSON.hasOwnProperty('modules'))) {
        errMsgArr.push(i18next.t('pluginmanifestrule_missingmodules'));
        return;
      }

      manifestValidations.modulesCheckList(manifestJSON);
      manifestValidations.configCheckList(manifestJSON);

    }

  },

  modulesCheckList: function (appJSON) {
    checkList = CONST.MANIFEST_CHECKLIST;

    for (var key in checkList) {
      var dataArray = manifestValidations.getJSONArrToValidate(key, checkList, appJSON);
      var dataArrayLength = dataArray.length;
      var keyLength = checkList[key].length;

      for (var i = 0; i < dataArrayLength; i++) {

        var data = dataArray[i];

        for (var j = 0; j < keyLength; j++) {

          var keyName = checkList[key][j];

          if (!data.hasOwnProperty(keyName)) {
            errMsgArr.push('[' + key.toUpperCase() + '] : ' + keyName + i18next.t('pluginmanifestrule_missingproperty'));
          }
          if (data[keyName] === '') {
            errMsgArr.push('[' + key.toUpperCase() + '] : ' + keyName + i18next.t('pluginmanifestrule_cantbeempty'));
          }

        }
      }
    }

    if (errMsgArr.length > 0) {
      return errMsgArr;
    }
  },

  configCheckList: function (appJSON) {

    var configArray = appJSON.config;

    for (var i = 0; i < configArray.length; i++) {

      var config = configArray[i];
      var configInputType, configInputOptions;

      if (!config.hasOwnProperty('name')) {
        errMsgArr.push(i18next.t('pluginmanifestrule_name_property_index') + (i + 1) + '.\n'); continue;

      } else if (config.name === '') {
        errMsgArr.push(i18next.t('pluginmanifestrule_property_empty_msg') + (i + 1) + '.\n'); continue;
      }

      if (config.hasOwnProperty('userdefined') && typeof config.userdefined !== 'boolean') {
        errMsgArr.push(i18next.t('pluginmanifestrule_property_booleantype_msg') + config.name + '.\n');
        continue;

      }

      if (config.userdefined === true) {

        if (!config.hasOwnProperty('type')) {
          errMsgArr.push(i18next.t('pluginmanifestrule_missingtype') + config.name + '.\n');

        } else if (config.type === '') {
          errMsgArr.push(i18next.t('pluginmanifestrule_emptytype') + config.name + '.\n');
        }

        configInputType = config.type;
        if (configInputType === 'select' || configInputType === 'checkbox') {
          if (!config.hasOwnProperty('options')) {
            errMsgArr.push(i18next.t('pluginmanifestrule_missingoptions') + config.name + '.\n');
          } else if (Array.isArray(config.options) && config.options.length < 1) {
            errMsgArr.push(i18next.t('pluginmanifestrule_emptyoptions') + config.name + '.\n');
          }
        }

        if (!config.default) {
          errMsgArr.push(i18next.t('pluginmanifestrule_emptydefault') + config.name + '.\n');
        }

        // Mandatory & Secure check
        if (!config.hasOwnProperty('mandatory')) {
          errMsgArr.push(i18next.t('pluginmanifestrule_mandatory_missing_msg') + config.name + '.\n');
        } else if (typeof config.mandatory !== 'boolean') {
          errMsgArr.push(i18next.t('pluginmanifestrule_mandatory_booleantype_msg') + config.name + '.\n');
        }

        if (!config.hasOwnProperty('secure')) {
          errMsgArr.push(i18next.t('pluginmanifestrule_secure_missing_msg') + config.name + '.\n');
        } else if (typeof config.secure !== 'boolean') {
          errMsgArr.push(i18next.t('pluginmanifestrule_secure_booleantype_msg') + config.name + '.\n');
        }

      } else {

        if (!config.hasOwnProperty('value')) {
          errMsgArr.push(i18next.t('pluginmanifestrule_missingvalue') + config.name + '.\n');
        } else if (config.value === '') {
          errMsgArr.push(i18next.t('pluginmanifestrule_emptyvalue') + config.name + '.\n');
        }
      }
    }

  },

  getJSONArrToValidate: function (key, checkList, appJSON) {
    var jsonArrToValidate = [];

    if (appJSON.modules && appJSON.modules.hasOwnProperty(key)) {
      jsonArrToValidate = appJSON.modules[key];
    } else {
      errMsgArr.push(key + i18next.t("pluginmanifestrule_missingmodule"));
    }

    return jsonArrToValidate;
  }
};

module.exports = {

  name: i18next.t('pluginmanifestrule_name'),
  run: function (projectRootDir, fileTree) {
    manifestValidations.validate(projectRootDir);
    return errMsgArr;
  }

};
