// zoho-extension-toolkit/lib/rules/PROJECTS.PluginManifestRule.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/

var fs = require('fs');
var path = require('path');
var i18next = require('i18next');
var CONST = require('../commands/constants.js');
var util = require('../utils');
var service = require('./../utils/service.js');
var isString = util.isString;
var isUndefined = util.isUndefined;
var isObject = util.isObject;
var isArray = util.isArray;
var isBoolean = util.isBoolean;

var SERVICE_NAME = CONST.SERVICE.PROJECTS; // constant
var MANIFEST_KEY = [
    "locale", "storage", "service", "modules", "whiteListedDomains"
];
var LOCATION_CONSTANT = [
    "app_settings", "task_tab", "issue_tab", "taskdetails_rightpanel", "issuedetails_rightpanel", 
    "top_band", "project_tab", "attachment_picker", "milestone_tab", "projectdetails_rightpanel"
];

function validatePluginManifest(projectRootDir) {
    var errors = [];
    var pluginManifest = require(path.join(projectRootDir, CONST.PLUGIN_MANIFEST));

    let manifestKeys = Object.keys(pluginManifest);

    //plugin manifest mandatory key validation
    for (var i = 0; i < MANIFEST_KEY.length; i++) {
        if (manifestKeys.indexOf(MANIFEST_KEY[i]) === -1) {
            errors.push("\"" + MANIFEST_KEY[i] + "\"" + i18next.t('pluginmanifest_keymissing'));
        }
    }

    //locale file exists validation
    var manifestLocale = pluginManifest.locale;
    if (manifestLocale) {
        if( isArray(manifestLocale) ) {
            var locales = manifestLocale;
            for( var i=0;i<locales.length; i++) {
                let _locale = locales[i];
                let relativePath = path.join( 'app', 'translations', _locale + '.json');
                let localePath = path.join(projectRootDir, relativePath);
                if (!fs.existsSync(localePath)) {
                    errors.push(i18next.t('desk_pluginmanifest_localenotfound'));
                }
            }
        } else {
            errors.push(i18next.t('desk_pluginmanifest_locale_invalidtype'));
        }
    }

    //cspDomains exists validation
    if (pluginManifest.cspDomains) {
        let cspDomains = pluginManifest.cspDomains;
        errors = errors.concat(service.validateCspDomains(cspDomains));
    }

    //storage type check
    if (typeof pluginManifest.storage !== 'boolean' ) {
        errors.push(i18next.t('desk_pluginmanifest_invalidstorage'));
    }

    //type validation
    var manifestType = pluginManifest.type;
    if (manifestType !== undefined) {
        if (!isString(manifestType) || ( isString(manifestType) && manifestType.trim() === "" ) ||
            (manifestType !== 'personal' && manifestType !== 'org')) {
            errors.push(i18next.t('desk_pluginmanifest_invalidtype'));
        }
    }
    
    // whitelisted domain validation
    var whitelisted = pluginManifest.whiteListedDomains;
    if (whitelisted) {
        if (!isArray(whitelisted)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidwhitelisteddomaintype'));
        } else {
            if (validateWhiteListedDomains(whitelisted)) {
                errors.push(i18next.t('desk_pluginmanifest_invalidwhitelisteddomain'));
            }
        }
    }

    // Modules validation
    var modules = pluginManifest.modules;
    if (!isUndefined(modules)) {
        if (!isObject(modules)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidmodules'));
        } else if (isUndefined(modules.widgets)) {
            errors.push(i18next.t('desk_pluginmanifest_widgetkeymissing'));
        } else if (!isArray(modules.widgets)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidwidgettype'));
        } else if (isArray(modules.widgets) && modules.widgets.length === 0) {
            errors.push(i18next.t('desk_pluginmanifest_widgetisempty'));
        } else {
            errors = errors.concat(validateWidgetConfigs(projectRootDir, modules.widgets));
        }
    }

    // Connectors validation
    var connectorsConfig = pluginManifest.connectors;
    if (!isUndefined(connectorsConfig)) {
        if (isArray(connectorsConfig)) {
            if(connectorsConfig.length > 5) {
                errors.push(i18next.t('projects_pluginmanifest_invalidconnectorcount'));
            }else {
                errors = errors.concat(validateConnectorConfig(connectorsConfig));
            }
        } else {
            errors.push(i18next.t('desk_pluginmanifest_invalidconnectortype'));
        }
    }

    // 'config' validations
    var configs = pluginManifest.config;
    if (!isUndefined(configs)) {
        if (isArray(configs)) {
            if (configs.length > 10) {
                errors.push(i18next.t('projects_pluginmanifest_configlimitexceeded'));
            } else {
                errors = errors.concat(validateConfigOptions(configs));
            }
        } else {
            errors.push(i18next.t('desk_pluginmanifest_invalidconfigtype'));
        }
    }

    return errors;
}

function validateWhiteListedDomains(domainsList) {
    
    var invalidURLs = domainsList.filter(function (domain) {
        return !util.isValidURL(domain);
    });

    return invalidURLs && invalidURLs.length > 0;
}

function validateWidgetConfigs(projectRootDir, widgets) {
    var errs = [], nameArray = [];

    for (var i = 0; i < widgets.length; i++) {
        var widget = widgets[i];
        var name = widget.name;
        if (isUndefined(name)) {
            errs.push(i18next.t('desk_pluginmanifest_missingwidgetname'));
        } else {
            if ( !isString(name) || ( isString(name) && name.trim() === "" ) ) {
                errs.push(i18next.t('desk_pluginmanifest_invalidwidgetname'));
            }
            if (!/^[a-zA-Z][a-zA-Z0-9-_ ]*$/i.test(name)) {
                errs.push(i18next.t("desk_pluginmanifest_invalidcharacters"));
            } 
        }

        var location = widget.location;
        var url = widget.url;
        var icon = widget.icon;

        if (isUndefined(location)) {
            errs.push(i18next.t('desk_pluginmanifest_missinglocationinwidget'));
        } else if (LOCATION_CONSTANT.indexOf(location) === -1) {
            errs.push(i18next.t('desk_pluginmanifest_invalidlocation'));
        } else if (nameArray.indexOf(location) >= 0) {
            errs.push(i18next.t("projects_pluginmanifest_duplicatelocation"));
        } else {
            nameArray.push(location);
        }

        if (isUndefined(url)) {
            errs.push(i18next.t('desk_pluginmanifest_missingurlinwidget'));
        } else if (!util.isValidURL(url)) {
            if (isString(url)) {
                var startsWithSlash = url.indexOf('/') === 0;
                if (startsWithSlash) {
                    if (!fs.existsSync(path.join(projectRootDir, url))) {
                        errs.push(i18next.t('desk_pluginmanifest_missingfileinappsource'));
                    }
                } else {
                    errs.push(i18next.t('desk_pluginmanifest_invalidurl'));
                }
            } else {
                errs.push(i18next.t('desk_pluginmanifest_invalidurltype'));
            }
        }

        if(location === "attachment_picker"){
            var err = validateIconForWidgetLocation(icon,projectRootDir,name);
            if(err!=""){
                errs.push(err);
            }
        }

        if(location === "projectdetails_rightpanel"){
            var err = validateIconForWidgetLocation(icon,projectRootDir,name);
            if(err!=""){
                errs.push(err);
            }
        }
    }

    return errs;
}

function validateIconForWidgetLocation(icon,projectRootDir,name) {
    if(isUndefined(icon)) {
        return i18next.t('desk_pluginmanifest_missingicon')+name;
    } else if(!isUndefined(icon)) {
        if(!isString(icon)) {
            return i18next.t('project_pluginmanifest_invalidicon');    
        } else if(icon.trim() != "") {
            var startsWithSlash = icon.indexOf('/') === 0;
            if (startsWithSlash) {
                if(!fs.existsSync(path.join(projectRootDir, icon))) {
                    return i18next.t('project_pluginmanifest_invalidiconforattachmentpicker');    
                }
            }
            else if(!util.isValidURL(icon)){
                return i18next.t('project_pluginmanifest_invalidurlforicon')+name;
            } 
        }
        else{
            return i18next.t('desk_pluginmanifest_missingicon')+name;
        }
    }
    return "";
}


function validateConnectorConfig(connectors) {
    var errors = [], nameArray = [];

    for (var i = 0; i < connectors.length; i++) {
        var connector = connectors[i];

        if (!isObject(connector)) {
            errors.push(i18next.t('desk_manifest_connectors_invalidtype'));
            continue;
        }
        if (isObject(connector) && Object.keys(connector).length === 0) {
            errors.push(i18next.t('desk_manifest_connectors_emptyobject'));
            continue;
        }

        var name = connector.connectionName;
        var service = connector.serviceName;
        var connectionLink = connector.connectionLinkName;
        var scope = connector.scope;
        var userAccess = connector.userAccess;
        var isUserDefinedService = connector.isUserDefinedService;

        if (isUndefined(name)) {
            errors.push(i18next.t('desk_manifest_connectors_name_missing'));
        } else if (!isString(name) || ( isString(name) && name.trim() === "") )  {
            errors.push(i18next.t('desk_manifest_connectors_name_empty'));
        } else {

            // unique 'name' validation
            if (!/^[a-zA-Z][a-zA-Z0-9-_ ]*$/i.test(name)) {
                errors.push(i18next.t('desk_manifest_connectors_name_invalid'));
            } 

            // Servicename property check
            if (isUndefined(service)) {
                errors.push(i18next.t('desk_manifest_connectors_service_missing'));
            } else if (!isString(service) || ( isString(service) && service.trim() === '') ) {
                errors.push(i18next.t('desk_manifest_connectors_service_invalid'));
            }

            // ConnectionLinkname property check
            if (isUndefined(connectionLink)) {
                errors.push(i18next.t('desk_manifest_connectors_connectionLink_missing'));
            } else if (!isString(connectionLink) || connectionLink.trim() === '') {
                errors.push(i18next.t('desk_manifest_connectors_connectionLink_invalid'));
            } else if (nameArray.indexOf(connectionLink.trim()) >= 0) {
                errors.push(i18next.t('projects_manifest_connectors_name_exists'));
            }
            nameArray.push(name.trim());

            if (isUndefined(userAccess)) {
                errors.push(i18next.t('projects_manifest_connectors_useraccess_missing'));
            } else if (!isBoolean(userAccess)) {
                errors.push(i18next.t('projects_manifest_connectors_useraccess_invalid'));
            }
           
            if (isUndefined(isUserDefinedService)) {
                errors.push(i18next.t('projects_manifest_connectors_userdefined_missing'));
            } else if (!isBoolean(isUserDefinedService)) {
                errors.push(i18next.t('projects_manifest_connectors_userdefined_invalid'));
            }

            if (!isUndefined(scope)) {
                if (isArray(scope)) {
                    for (var j = 0; j < scope.length; j++) {
                        if (!isString(scope[j])) {
                            errors.push(i18next.t('desk_manifest_connectors_scope_invalid'));
                            break;
                        }
                        if(scope[j].trim()===''){
                            errors.push(i18next.t('desk_manifest_connectors_scope_empty'));
                        }
                    }
                } else {
                    errors.push(i18next.t('desk_manifest_connectors_scope_invalidtype'));
                }
            }
        }
    }
    return errors;
}

function validateConfigOptions(configs) {
    var errors = [], nameArray = [];

    for (var i = 0; i < configs.length; i++) {
        var config = configs[i];
        if (!isObject(config)) {
            errors.push(i18next.t('projects_pluginmanifest_invalidconfigtype'));
            continue;
        }
        if( isObject(config) && Object.keys(config).length === 0 ) {
            errors.push(i18next.t('desk_pluginmanifest_emptyconfig'));
            continue;
        }

        var name = config.name;
        //var defaultvalue = config.default;
        var type = config.type;
        var userdefined = config.description;
        var mandatory = config.is_mandatory;
        var secure = config.is_secure;
        var defaultconfig = config.default;

        if (isUndefined(name)) {
            errors.push(i18next.t('desk_pluginmanifest_missingconfigname'));
        } else {
            if (!isString(name) || ( isString(name) && name.trim() === "" ) ) {
                errors.push(i18next.t('desk_pluginmanifest_invalidconfigname'));
            }
            // unique 'name' validation
            if (!/^[a-zA-Z][a-zA-Z0-9-_ ]*$/i.test(name)) {
                errors.push(i18next.t("desk_pluginmanifest_invalidcharacters"));
            } else if (isString(name) && nameArray.indexOf(name.trim()) >= 0) {
                errors.push(i18next.t("desk_pluginmanifest_duplicatename"));
            } else {
                nameArray.push(name.trim());
            }
        }

        var undefinedList = [type, mandatory, secure].filter(isUndefined);
        var configType = ["text", "password", "checkbox", "url", "number", "multiline", "selectbox", "hidden"];
        if (undefinedList.length > 0 && undefinedList.length < 4) {
            errors.push(i18next.t('projects_pluginmanifest_requiredconfigproperties'));
            continue;
        }
        if (undefinedList.length === 0) {
            if (!isString(type)) {
                errors.push(i18next.t('projects_pluginmanifest_invalidtypeinconfig'));
            } else if (configType.indexOf(type) === -1){
                errors.push(i18next.t('projects_pluginmanifest_invalidlistoftypeinconfig'));
            } else if(type === "selectbox") {
                var options = config.options;
                if(isUndefined(options)){
                    errors.push(i18next.t('projects_pluginmanifest_missingoptioninconfig'));
                } else if(!isArray(options)) {
                    errors.push(i18next.t('projects_pluginmanifest_invalidoptioninconfig'));
                }
            }
            if (!isBoolean(mandatory)) {
                errors.push(i18next.t('projects_pluginmanifest_invalidmandatorytype'));
            }
            if (!isBoolean(secure)) {
                errors.push(i18next.t('projects_pluginmanifest_invalidsecuretype'));
            }
            if (!isUndefined(defaultconfig)) { 
                if(!isString(defaultconfig)) {
                    errors.push(i18next.t('projects_pluginmanifest_invaliddefaulttype'));
                } else if(configType.indexOf(type) != -1){
                    if(type === "checkbox" && !isBoolean(defaultconfig)) {
                        errors.push(i18next.t('projects_pluginmanifest_invaliddefaulttypeforcheckbox'));
                    } else if(type === "url" && !isValidURL(defaultconfig)) {
                        errors.push(i18next.t('projects_pluginmanifest_invaliddefaulttypeforurl'));
                    } else if(type === "number" && isNaN(defaultconfig)) {
                        errors.push(i18next.t('projects_pluginmanifest_invaliddefaulttypefornumber'));
                    } else if(type === "selectbox") {
                        var options = config.options;
                        if(!isUndefined(options) && isArray(options) && options.indexOf(defaultconfig) != -1){
                            errors.push(i18next.t('projects_pluginmanifest_invaliddefaulttypeforselectbox'));
                        }
                    }
                }
            }
        }
    }
    return errors;
}

module.exports = {

    name: i18next.t('pluginmanifestrule_name'),
    run: function (projectRootDir, fileTree) {
        return validatePluginManifest(projectRootDir);
    }
}
