// zoho-extension-toolkit/lib/rules/SHOW.PluginManifestRule.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/

var path = require('path');
var fs = require('fs');
var i18next = require('i18next');
var util = require('../utils');
var CONST = require('../commands/constants');
var service = require('./../utils/service.js');
var isUndefined = util.isUndefined;
var isBoolean = util.isBoolean;
var isObject = util.isObject;
var isString = util.isString;
var isArray = util.isArray;
var isAllowedString = util.isAllowedString;
var isValidURL = util.isValidURL;

var LOCATION_CONSTANT = [
    "show.insert.text", "show.insert.shape", "show.insert.picture", 
    "show.insert.media","show.insert.addon"
];
var MANIFEST_KEY = [
    "locale", "storage", "service",
     "modules", "connectors"
   
];

function validatePluginManifest(projectRootDir) {
    var errors = [];

    var manifest = require(path.join(projectRootDir, CONST.PLUGIN_MANIFEST));

    let manifestKeys = Object.keys(manifest);

    //plugin manifest mandatory key validation
    for (var i = 0; i < MANIFEST_KEY.length; i++) {
        if (manifestKeys.indexOf(MANIFEST_KEY[i]) === -1) {
            errors.push("\"" + MANIFEST_KEY[i] + "\"" + i18next.t('pluginmanifest_keymissing'));
        }
    }
    if( errors.length > 0 ) return errors;

    //locale file exists validation
    if (manifest.locale) {
        if( isArray(manifest.locale) ) {
            var locales = manifest.locale;
            for( var i=0;i<locales.length; i++) {
                let _locale = locales[i];
                let relativePath = path.join( 'app', 'translations', _locale + '.json');
                let localePath = path.join(projectRootDir, relativePath);
                if (!fs.existsSync(localePath)) {
                    errors.push(i18next.t('desk_pluginmanifest_localenotfound') );
                }
            }
        } else {
            errors.push(i18next.t('desk_pluginmanifest_locale_invalidtype'));
        }
    }

    if (typeof manifest.storage !== 'boolean' ) {
        errors.push(i18next.t('desk_pluginmanifest_invalidstorage'));
    }

    if (manifest.type !== undefined) {
        if (!isString(manifest.type) || ( isString(manifest.type) && manifest.type.trim() === "" ) ||
            (manifest.type !== 'personal' && manifest.type !== 'org')) {
            errors.push(i18next.t('desk_pluginmanifest_invalidtype'));
        }
    }
    
    if (!isUndefined(manifest.modules)) {
        if (!isObject(manifest.modules)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidmodules'));
        } else if (isUndefined(manifest.modules.widgets)) {
            errors.push(i18next.t('desk_pluginmanifest_widgetkeymissing'));
        } else if (!isArray(manifest.modules.widgets)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidwidgettype'));
        } else if (isArray(manifest.modules.widgets) && manifest.modules.widgets.length === 0) {
            errors.push(i18next.t('desk_pluginmanifest_widgetisempty'));
        } else {
            errors = errors.concat(validateWidgetConfigs(projectRootDir, manifest.modules.widgets));
        }
    }

     if (manifest.whiteListedDomains) {
        if (!isArray(manifest.whiteListedDomains)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidwhitelisteddomaintype'));
        } else {
            if (validateWhiteListedDomains(manifest.whiteListedDomains)) {
                errors.push(i18next.t('desk_pluginmanifest_invalidwhitelisteddomain'));
            }
        }
    } 

    // 'connectors' validation
    var connectorsConfig = manifest.connectors;
    if (!isUndefined(connectorsConfig)) {
        if (isArray(connectorsConfig)) {
            errors = errors.concat(validateConnectorConfig(connectorsConfig));
        } else {
            errors.push(i18next.t('desk_pluginmanifest_invalidconnectortype'));
        }
    }

    // 'deskOAuth' validation
    var deskOAuthConfig = manifest.deskOauth;
    if (!isUndefined(deskOAuthConfig)) {
        if (isObject(deskOAuthConfig)) {
            errors = errors.concat(validateDeskOAuthConfig(deskOAuthConfig));
        } else {
            errors.push(i18next.t('desk_manifest_doc_invalidtype'));
        }
    } 

    // 'TPAOAuth' validation
     var tpaOAuth = manifest.TPAOauth;
    if( !isUndefined(tpaOAuth) ) {
        if(isArray(tpaOAuth)) {
            errors = errors.concat(validateTPAOAuthConfig(tpaOAuth));
        } else {
            errors.push(i18next.t('desk_manifest_tpo_invalidtype'));
        }
    } 

    // 'callbackListeners' validation
    var callbackListener = manifest.callbackListener;
    if( !isUndefined(callbackListener) && isObject(callbackListener) ) {
        var callbackListenerKeys = [
            'onDeskAuthorise', 'onTPAAuthorise', 'onTPARevoke', 
            'onUpdate', 'onConfigParamChange', 'onUninstall' 
        ];
        for( var i=0;i<callbackListenerKeys.length;i++) {
            var cbKey = callbackListenerKeys[i];
            if( !isUndefined(callbackListener[cbKey]) && !isValidURL(callbackListener[cbKey]) ) {
                errors.push('Invalid URL value provided in manifest for key: '+ cbKey);
            }
        }
    } else if ( !isUndefined(callbackListener) && !isObject(callbackListener) ) {
        errors.push(i18next.t('desk_pluginmanifest_callbacklistener_invalidtype'));
    } 

    // 'config' validations
     var configs = manifest.config;
    if (!isUndefined(configs)) {
        if (isArray(configs)) {
            if (configs.length > 20) {
                errors.push(i18next.t('desk_pluginmanifest_configlimitexceeded'));
            } else {
                errors = errors.concat(validateConfigOptions(configs));
            }
        } else {
            errors.push(i18next.t('desk_pluginmanifest_invalidconfigtype'));
        }
    } 

    return errors;
}

function validateTPAOAuthConfig(configs) {
    var errors = [], nameArray = [];

    for (var i = 0; i < configs.length; i++) {
        var config = configs[i];
        if (!isObject(config)) {
            errors.push(i18next.t('desk_manifest_tpo_config_invalidtype'));
            continue;
        }
        if( isObject(config) && Object.keys(config).length === 0 ) {
            errors.push(i18next.t('desk_manifest_tpo_config_empty'));
            continue;
        }

        var name = config.name;
        var clientID = config.clientId;
        var clientSecret = config.clientSecret;
        var authoriseURL = config.authoriseUrl;
        var tokenURL = config.tokenUrl;
        var revokeURL = config.revokeUrl;
        var params = config.params;

        if (isUndefined(name)) {
            errors.push(i18next.t('desk_manifest_tpo_name_missing'));
        } else {
            // unique 'name' validation
            if( !isString(name) || name === '' ) {
                errors.push(i18next.t('desk_manifest_tpo_name_empty'));
            } else if (!isAllowedString(name)) {
                errors.push(i18next.t('desk_manifest_tpo_name_invalidtype'));
            } else if (nameArray.indexOf(name.trim()) >= 0) {
                errors.push(i18next.t("desk_manifest_tpo_name_alreadyexists"));
            } else {
                nameArray.push(name.trim());
            }
        }


        if( isUndefined(clientID) ) {
            errors.push(i18next.t('desk_manifest_tpo_clientid_missing'));
        } else if( clientID === "" ) {
            errors.push(i18next.t('desk_manifest_tpo_clientid_invalidtype'));
        }

        if( isUndefined(clientSecret) ) {
            errors.push(i18next.t('desk_manifest_tpo_clientsecret_missing'));
        } else if( clientSecret === "" ) {
            errors.push(i18next.t('desk_manifest_tpo_clientsecret_invalidtype'));
        }

        if( isUndefined(authoriseURL) ) {
            errors.push(i18next.t('desk_manifest_tpo_authoriseurl_missing'));
        } else if( authoriseURL === "" || !isValidURL(authoriseURL)) {
            errors.push(i18next.t('desk_manifest_tpo_authoriseurl_invalidurl'));
        }

        if( isUndefined(tokenURL) ) {
            errors.push(i18next.t('desk_manifest_tpo_tokenurl_missing'));
        } else if( tokenURL === "" || !isValidURL(tokenURL)) {
            errors.push(i18next.t('desk_manifest_tpo_tokenurl_invalidurl'));
        }

        if( !isUndefined(revokeURL) && !isValidURL(revokeURL)) {
            errors.push(i18next.t('desk_manifest_tpo_revokeurl_invalidurl'));
        }

        if( !isUndefined(params) && !isObject(params) ) {
            errors.push(i18next.t('desk_manifest_tpo_params_invalidtype'));
        }
    }

    return errors;
}
function validateDeskOAuthConfig(config) {
    var errors = [];

    var clientID = config.clientId;
    var name = config.name;
    var clientSecret = config.clientSecret;
    var scopes = config.scopes;
    var params = config.params;

    if( !clientID ) {
        errors.push(i18next.t('desk_manifest_doc_clientid_missing'));
    } else if( !isString(clientID) )  {
        errors.push(i18next.t('desk_manifest_doc_clientid_invalidtype'));
    }

    if( !name ) {
        errors.push(i18next.t('desk_manifest_doc_name_missing'));
    } else if( !isString(name)) {
        errors.push(i18next.t('desk_manifest_doc_name_invalidtype'));
    }

    if( !clientSecret ) {
        errors.push(i18next.t('desk_manifest_doc_clientsecret_missing'));
    } else if( !isString(clientSecret) ) {
        errors.push(i18next.t('desk_manifest_doc_clientsecret_invalidtype'));
    }

    if(scopes && !Array.isArray(scopes)) {
        errors.push(i18next.t('desk_manifest_doc_scopes_invalidtype'));
    }
    if(params && !isObject(params)) {
        errors.push(i18next.t('desk_manifest_doc_params_invalidtype'));
    }

    return errors;
}
function validateConfigOptions(configs) {
    var errors = [], nameArray = [];

    for (var i = 0; i < configs.length; i++) {
        var config = configs[i];
        if (!isObject(config)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidconfigtype'));
            continue;
        }
        if( isObject(config) && Object.keys(config).length === 0 ) {
            errors.push(i18next.t('desk_pluginmanifest_emptyconfig'));
            continue;
        }

        var name = config.name;
        var defaultvalue = config.default;
        var type = config.type;
        var userdefined = config.userdefined;
        var mandatory = config.mandatory;
        var secure = config.secure;

        if (isUndefined(name)) {
            errors.push(i18next.t('desk_pluginmanifest_missingconfigname'));
        } else {
            if (!isString(name) || ( isString(name) && name.trim() === "" ) ) {
                errors.push(i18next.t('desk_pluginmanifest_invalidconfigname'));
            }
            // unique 'name' validation
            if (!/^[a-zA-Z][a-zA-Z0-9-_ ]*$/i.test(name)) {
                errors.push(i18next.t("desk_pluginmanifest_invalidcharacters") + "'config' : " + name);
            } else if (isString(name) && nameArray.indexOf(name.trim()) >= 0) {
                errors.push(i18next.t("desk_pluginmanifest_duplicatename") + "'config' : " + name);
            } else {
                nameArray.push(name.trim());
            }
        }

        var undefinedList = [type, userdefined, mandatory, secure].filter(isUndefined);
        var configType = ["text", "checkbox", "number", "textarea", "picklist", "radio", "url", "email"];
        if (undefinedList.length > 0 && undefinedList.length < 4) {
            errors.push(i18next.t('desk_pluginmanifest_requiredconfigproperties') + name);
            continue;
        }
        if (undefinedList.length === 0) {
            if (!isString(type)) {
                errors.push(i18next.t('projects_pluginmanifest_invalidtypeinconfig'));
            } else if (configType.indexOf(type) === -1){
                errors.push(i18next.t('desk_pluginmanifest_invalidlistoftypeinconfig'));
            }
            if (!isBoolean(userdefined)) {
                errors.push(i18next.t('desk_pluginmanifest_invaliduserdefinedconfigtype'));
            }
            if (!isBoolean(mandatory)) {
                errors.push(i18next.t('desk_pluginmanifest_invalidmandatorytype'));
            }
            if (!isBoolean(secure)) {
                errors.push(i18next.t('desk_pluginmanifest_invalidsecuretype'));
            }
        }
    }
    return errors;
}

function validateConnectorConfig(connectors) {
    var errors = [], nameArray = [];

    for (var i = 0; i < connectors.length; i++) {
        var connector = connectors[i];

        if (!isObject(connector)) {
            errors.push(i18next.t('desk_pluginmanifest_invalidconnectorstype'));
            continue;
        }
        if (isObject(connector) && Object.keys(connector).length === 0) {
            errors.push(i18next.t('desk_manifest_connectors_emptyobject'));
            continue;
        }

        var name = connector.connectionName;
        var sharedBY = connector.sharedBy;
        var scope = connector.scope;

        if (isUndefined(name)) {
            errors.push(i18next.t('desk_pluginmanifest_missingconnectorname'));
        } else if (!isString(name) || ( isString(name) && name.trim() === "") )  {
            errors.push(i18next.t('desk_pluginmanifest_invalidconnectorname'));
        } else {

            // unique 'name' validation
            if (!/^[a-zA-Z][a-zA-Z0-9-_ ]*$/i.test(name)) {
                errors.push(i18next.t("desk_pluginmanifest_invalidcharacters") + "'connectors' : " + name);
            } else if (nameArray.indexOf(name.trim()) >= 0) {
                errors.push(i18next.t("desk_pluginmanifest_duplicatename") + "'connectors' : " + name);
            }
            nameArray.push(name.trim());

            // shared By property check
            if (isUndefined(sharedBY)) {
                errors.push(i18next.t('show_pluginmanifest_missingconnectorsharedby') + name);
            } else if (!Number.isInteger(sharedBY))  {
                errors.push(i18next.t('show_pluginmanifest_invalidconnectorsharedby') + name);
            }

            // ConnectionLinkname property check
           /*  if (isUndefined(connectionLink)) {
                errors.push(i18next.t('desk_manifest_connectors_connectionLink_missing'));
            } else if (!isString(connectionLink)) {
                errors.push(i18next.t('desk_manifest_connectors_connectionLink_invalid'));
            } */

            if (!isUndefined(scope)) {
                if (isArray(scope)) {
                    for (var j = 0; j < scope.length; j++) {
                        if (!isString(scope[j])) {
                            errors.push(i18next.t('desk_pluginmanifest_invalidscopearraytype') + name);
                            break;
                        }
                        if(scope[j].trim()===''){
                            errors.push(i18next.t('desk_pluginmanifest_scopeisempty'));
                        }
                    }
                } else {
                    errors.push(i18next.t('desk_pluginmanifest_invalidconnectorscope') + name);
                }
            }
        }
    }
    return errors;
}

function validateWhiteListedDomains(domainsList) {
    var invalidURLs = domainsList.filter(function (domain) {
        return !util.isValidURL(domain);
    });

    return invalidURLs && invalidURLs.length > 0;
}

function validateWidgetConfigs(projectRootDir, widgets) {
    var errs = [], nameArray = [], bgLocationFlag = false;

    for (var i = 0; i < widgets.length; i++) {
        var widget = widgets[i];
        var name = widget.name;
        if (isUndefined(name)) {
            errs.push(i18next.t('desk_pluginmanifest_missingwidgetname') + name);
        } else {
            if ( !isString(name) || ( isString(name) && name.trim() === "" ) ) {
                errs.push(i18next.t('desk_pluginmanifest_invalidwidgetname') + name);
            }
            if (!/^[a-zA-Z][a-zA-Z0-9-_ ]*$/i.test(name)) {
                errs.push(i18next.t("desk_pluginmanifest_invalidcharacters") + "'widgets' : " + name);
            } else if (isString(name) && nameArray.indexOf(name.trim()) >= 0) {
                errs.push(i18next.t("desk_pluginmanifest_duplicatename") + "'widgets' : " + name);
            } else if( isString(name) ) {
                nameArray.push(name.trim());
            }
        }

        var location = widget.location;
        var url = widget.url;
        var logo = widget.logo;
        var icon = widget.icon;
       

        if (isUndefined(location)) {
            errs.push(i18next.t('desk_pluginmanifest_missinglocationinwidget') + name);
        } else if (LOCATION_CONSTANT.indexOf(location) === -1) {
            errs.push(i18next.t('desk_pluginmanifest_invalidlocation') + name);
        } 

        if (isUndefined(url)) {
            errs.push(i18next.t('desk_pluginmanifest_missingurlinwidget') + name);
        } else if (!util.isValidURL(url)) {
            if (isString(url)) {
                var startsWithSlash = url.indexOf('/') === 0;
                if (startsWithSlash) {
                    if (!fs.existsSync(path.join(projectRootDir, url))) {
                        errs.push(i18next.t('desk_pluginmanifest_missingfileinappsource'));
                    }
                } else {
                    errs.push(i18next.t('desk_pluginmanifest_invalidurl') + name);
                }
            } else {
                errs.push(i18next.t('desk_pluginmanifest_invalidurltype') + name);
            }
        }

        let serviceName = service.getServiceName(projectRootDir);

        if (!isUndefined(logo)) {
            if (isString(logo)) {
                let extnName = path.extname(logo);
                if (CONST.ALLOWED_EXTNS[serviceName].images.indexOf(extnName) === -1) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidlogoextn'));
                }
                if (logo.indexOf('/') !== 0) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidlogo'));
                }
                if (!fs.existsSync(path.join(projectRootDir, logo))) {
                    errs.push(i18next.t('desk_pluginmanifest_missinglogofile'));
                }
            } else {
                errs.push(i18next.t('deskpluginmanifest_invalidlogotype'));
            }
        }

        if (!isUndefined(icon)) {
            if (isString(icon)) {
                let extnName = path.extname(icon);
                if (CONST.ALLOWED_EXTNS[serviceName].images.indexOf(extnName) === -1) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidiconextn'));
                }
                if (icon.indexOf('/') !== 0) {
                    errs.push(i18next.t('desk_pluginmanifest_invalidicon'));
                }
                if (!fs.existsSync(path.join(projectRootDir, icon))) {
                    errs.push(i18next.t('desk_pluginmanifest_missingiconfile'));
                }
            } else {
                errs.push(i18next.t('deskpluginmanifest_invalidicontype'));
            }
        }
    }
    return errs;
}

module.exports = {
    name: i18next.t('pluginmanifestrule_name'),
    run: function (projectRootDir, fileTree) {
        return validatePluginManifest(projectRootDir);
    }
};
