// zoho-extension-toolkit/lib/commands/cloud_run.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/
var chalk = require('chalk');
var i18next = require('i18next');
var service = require('../utils/service.js');
const _ = require('../util_modules');
const store = require('../configstore');
const Credential = require('../internal/credential');
const API = require('../internal/api');
const validate = require('./validation_api');
const open = require('open');
const font = require('ansi-colors');
const prompt = require('../prompt');
var CONST = require('./constants.js');

function getOrglist( app_data){
    return new Promise(function () {
        let tokenObj = store.get('credential', null);
        service.validateTokenObject(tokenObj);
        Credential.init(tokenObj, false);
        _.LOGGER.message(i18next.t('waiting_for_response'))  ;
        new API()
          .get('/workspace/restapi/'+store.get('default_workspace')+'/apps/orglist/'+app_data.appId, {
            origin: _.CONSTANTS.ORIGIN.sigma,
            dctype: store.get('dctype'),
            params: {
            for : 'run',
            zet_call : 'true',
            force_draft : 'true'
                    },
            auth: true,
            json: false
          })
          .then((res) => {
            var data = res.body;
            if(!data.orglist || data.orglist.length == 0){
                _.LOGGER.error(chalk.red(i18next.t('cloud_run_no_orgs')+data.service));
            }else{
                //many orgs found for the service
                if(data.orglist.length == 1){
                    runApp(data.orglist[0].id, app_data);
                }else{
                  var orglistdata = [];
                  data.orglist.forEach(function(obj) 
                  { 
                  orglistdata.push({
                      'name' : obj.display_name,
                      'value' : obj.id
                  });
                  });

                  prompt.ask(
                  prompt.question(
                      'scopeId',
                      i18next.t('cloud_run_select_org'),
                      {
                      type: 'list',
                      choices: orglistdata
                      }
                  )
                  )
                  .then((ans) => {
                     runApp(ans.scopeId, app_data);
                  })
                }
            }
                  
          });
      }); 
    }

    function runApp(scopeId, app_data){
        let tokenObj = store.get('credential', null);
        service.validateTokenObject(tokenObj);
        Credential.init(tokenObj, false);
        _.LOGGER.message(i18next.t('waiting_for_response'))  ;
        new API()
          .post('/workspace/restapi/'+store.get('default_workspace')+'/apps/run/'+app_data.appId, {
            origin: _.CONSTANTS.ORIGIN.sigma,
            dctype: store.get('dctype'),
            params: {
                scope_id : scopeId,
                validate : true,
                zet_call:true
                  },
            auth: true,
            json: false
          })
          .then((res) => {
            var data = res.body;
            _.LOGGER.info();
            _.LOGGER.info(i18next.t('cloud_run_view_url'));
            _.LOGGER.info();
            //Authorize URL will not be available for new run mode
            if(data.message == 'Authorized' || (!data.message && !data.authorize_url)){
              _.LOGGER.info(font.bold.underline(data.run_url));
              open(data.run_url);
            }else{
              _.LOGGER.info(font.bold.underline(data.authorize_url+'&zet_call=true'));
                open(data.authorize_url+'&zet_call=true');
            }
          });
    }

    function stopApp(app_data){
      let tokenObj = store.get('credential', null);
      service.validateTokenObject(tokenObj);
      Credential.init(tokenObj, false);
      _.LOGGER.message(i18next.t('waiting_for_response'))  ;
      new API()
        .post('/workspace/restapi/'+store.get('default_workspace')+'/apps/stop/'+app_data.appId, {
          origin: _.CONSTANTS.ORIGIN.sigma,
          dctype: store.get('dctype'),
          params: {
             
                  },
          auth: true,
          json: false
        })
        .then((res) => {
          var data = res.body;
          //console.log(data);
          if(data.success){
            _.LOGGER.success(i18next.t('cloud_stop_success'));
          }else{
            _.LOGGER.error(i18next.t('cloud_stop_failed'));
          }
        });
  }

module.exports = {
  run: function (projectRootDir) {
    return new Promise(function (resolve) {
        service.validateDefaultWorkspace(store.get('default_workspace'));
        service.validateAppInfoFile(service.isJsonFilePresent(projectRootDir));
        service.readJsonFromFile(projectRootDir).then((app_data) => {
        //console.log(app_data);
          service.validateJSONData(app_data,store.get('default_workspace'));
          validate.run(projectRootDir).then(()=>{
           require('./status').compare(projectRootDir, true).then( () => {  
              var serviceName = service.getServiceName(projectRootDir);
              if(serviceName === CONST.SERVICE.MAIL){
                runApp(-1, app_data);
              }else{
                getOrglist( app_data).then(()=>{
                  resolve();
                });
              }
            });  
          });
      });
    });
  },

  stop: function (projectRootDir) {
        service.validateDefaultWorkspace(store.get('default_workspace'));
        service.validateAppInfoFile(service.isJsonFilePresent(projectRootDir));
        service.readJsonFromFile(projectRootDir).then((app_data) => {
        //console.log(app_data);
          service.validateJSONData(app_data,store.get('default_workspace'));
           stopApp( app_data);
      });
  }
};
