// zoho-extension-toolkit/lib/commands/constants.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/

const DEFAULT_EXTENSION_OBJ = {
  'images': [".jpg", ".jpeg", ".png", ".gif", ".svg"],
  'styles': [".css", ".woff", ".woff2", ".ttf", ".eot", ".webm"],
  'scripts': [".js"],
  'multimedia': [".mp3", ".mp4"], 
  'data': [".json",".txt",".md",".xml",".dre"],
  'views': [".html"],
  'translations': [".json"]
};
module.exports = {
  FILES_COUNT_LIMIT: 250,
  MAX_DIST_SIZE: 20000000,
  FILE_SIZE_LIMIT: 5000000,
  MAX_FILE_NAME_LEN: 50,
  MAX_FILE_NAME_LEN_CRM: 250,
  ALLOWED_EXTNS: {
    'ZOHO': DEFAULT_EXTENSION_OBJ,
    'CRM': DEFAULT_EXTENSION_OBJ,
    'CATALYST': DEFAULT_EXTENSION_OBJ,
    'DESK': DEFAULT_EXTENSION_OBJ,
    'PROJECTS': DEFAULT_EXTENSION_OBJ,
    'QNTRL': DEFAULT_EXTENSION_OBJ,
    'MAIL': DEFAULT_EXTENSION_OBJ,
    'CREATOR': DEFAULT_EXTENSION_OBJ,
    'TEAMINBOX' : DEFAULT_EXTENSION_OBJ,
    'SDP': DEFAULT_EXTENSION_OBJ,
    'FINANCE': DEFAULT_EXTENSION_OBJ,
    'IOT': DEFAULT_EXTENSION_OBJ,
    'INVENTORY': DEFAULT_EXTENSION_OBJ,
    'PEOPLE' : DEFAULT_EXTENSION_OBJ,
    'EXPENSE' : DEFAULT_EXTENSION_OBJ,
    'BIGIN' : DEFAULT_EXTENSION_OBJ,
    'BILLING' : DEFAULT_EXTENSION_OBJ,
    'RECRUIT' : DEFAULT_EXTENSION_OBJ,
    'FSM' : DEFAULT_EXTENSION_OBJ,
    'LOGS360CLOUD' : DEFAULT_EXTENSION_OBJ
  },
  MANIFEST_CHECKLIST: {
    'widgets': ['location', 'url'],
    'connectors': ['name', 'service']
  },
  PLUGIN_MANIFEST: 'plugin-manifest.json',
  RESOURCE_JSON: 'resources.json',
  SERVICE: {
    'CRM': 'CRM',
    'CATALYST': 'CATALYST',
    'DESK': 'DESK',
    'MAIL': 'MAIL',
    'PROJECTS': 'PROJECTS',
    'SDP': 'SDP',
    'FINANCE': 'FINANCE',
    'IOT': 'IOT',
    'QNTRL' : 'QNTRL',
    'SHOW' : 'SHOW',
    'CREATOR': 'CREATOR',
    'TEAMINBOX' : 'TEAMINBOX',
    'RECRUIT' : 'RECRUIT',
    'BIGIN' : 'BIGIN',
    'INVENTORY' : 'INVENTORY',
    'PEOPLE' : 'PEOPLE',
    'EXPENSE' : 'EXPENSE',
    'BILLING' : 'BILLING',
    'FSM' : 'FSM',
    'LOGS360CLOUD' : 'LOGS360CLOUD'
  },
  APPTYPE: {
    'Widgets': 1 ,
    'ZES': 2 ,
    'CRM': 3 ,
    'ZohoCRM': 4,
    'vCRM': 5 ,
    'ExternalExtensions': 6,
    'ChildWidgets': 7
  },
  SERVICE_APPTYPE_MAPPING:{
    //'PROJECTS': 'Widgets',
     'QNTRL' : 'Widgets',
    //'CRM': 'ExternalExtensions',
    // 'CATALYST': 'Widgets',
     'DESK': 'Widgets',
    'MAIL': 'Widgets'
    
    // 'SDP': 'Widgets',
    // 'FINANCE': 'ExternalExtensions',
    // 'IOT': 'Widgets',
    // 'SHOW' : 'Widgets',
    // 'CREATOR': 'Widgets'
  },
  //Service Name should be as in pulgin manifest
  SERVICE_NO: {
    'CRM': 2,
    'CATALYST': 10,
    'DESK': 3,
    'MAIL': 6,
    'PROJECTS': 4,
    'SDP': 8,
    'FINANCE': 11,
    'IOT': 9,
    'QNTRL' : 5,
    'SHOW' : 7,
    'CREATOR': 16
  },
  //ZET_API_MODEL_SERVICES : ['DESK']  removing DESK from zet online
  ZET_API_MODEL_SERVICES : ['MAIL'],
  APP_JSON : "app_info.json",
  ExpireNoOfDays : 1
};
