// zoho-extension-toolkit/lib/commands/create_app.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/
const { exec } = require('child_process');
var chalk = require('chalk');
var path = require('path');
var i18next = require('i18next');
var service = require('../utils/service.js');
var CONST = require('./constants.js');
const _ = require('../util_modules');
const store = require('../configstore');
const Credential = require('../internal/credential');
var fs = require('fs');
const API = require('../internal/api');
const validate = require('./validation_api');
const FormData = require('form-data');

function createApp(zipPath,serviceNo,apptypeNo,projectName,description,projectRootDir){
  return new Promise(function (resolve) {
    let tokenObj = store.get('credential', null);
    service.validateTokenObject(tokenObj);
    service.validateDefaultWorkspace(store.get('default_workspace'));
    
    Credential.init(tokenObj, false);

    const form = new FormData();
    form.append('app_zip', fs.createReadStream(zipPath), { filename: 'appZip.zip' } );
    _.LOGGER.message(i18next.t('waiting_for_response'))  ;
    new API()
      .post('/workspace/restapi/'+store.get('default_workspace')+'/apps', {
        origin: _.CONSTANTS.ORIGIN.sigma,
        dctype: store.get('dctype'),
        params: {
          service_no: serviceNo,
          app_name: projectName,
          app_type: apptypeNo,
          description: description
                },
        data: form,
        auth: true,
        json: false
      })
      .then((res) => {
                var data = res.body;
              // data.workspaceName = store.get('default_workspace');
                if(data.appObj){
                  var jsonfile = '{ "workspaceName":"'+store.get('default_workspace')+'","appUuid":"'+data.appObj.appUUID+'" , "appId":"'+data.appObj.appId+'"}';
                  service.writeJsonAsFile(JSON.parse(jsonfile), projectRootDir).then(()=>{
                    _.LOGGER.success(chalk.bold.green(i18next.t('create_app_success')));
                    _.LOGGER.info(chalk.bold.cyan(i18next.t('init_runcmd')));
                    _.LOGGER.info(chalk.bold.white(i18next.t('init_cd'))+ chalk.green('\''+projectName+'\''));
                    _.LOGGER.info(chalk.bold.white(i18next.t('init_run')));
                    resolve();
                  });
                }else{
                  _.LOGGER.error(chalk.red(i18next.t('create_app_failed')));
                }
                resolve();
                });
  });
}

function createAppWithoutZip(serviceNo,apptypeNo,projectName,description,projectPath){
  return new Promise(function (resolve) {
    let tokenObj = store.get('credential', null);
    service.validateTokenObject(tokenObj);
    service.validateDefaultWorkspace(store.get('default_workspace'));
    Credential.init(tokenObj, false);
    _.LOGGER.message(i18next.t('waiting_for_response'))  ;
    new API()
      .post('/workspace/restapi/'+store.get('default_workspace')+'/apps', {
        origin: _.CONSTANTS.ORIGIN.sigma,
        dctype: store.get('dctype'),
        params: {
          service_no: serviceNo,
          app_name: projectName,
          app_type: apptypeNo,
          description: description
                },
        auth: true,
        json: false
      })
      .then((res) => {
                var data = res.body;
                if(data.appObj){
                  var jsonfile = '{ "workspaceName":"'+store.get('default_workspace')+'","appUuid":"'+data.appObj.appUUID+'" , "appId":"'+data.appObj.appId+'"}';
                  service.writeJsonAsFile(JSON.parse(jsonfile), path.join(projectPath, projectName)).then(()=>{
                    _.LOGGER.success(chalk.bold.green(i18next.t('create_app_success')));
                    resolve();
                  });
                }else{
                  _.LOGGER.error(chalk.red(i18next.t('create_app_failed')));
                  resolve();
                }
               
                });
  });
}

//temp for desk only
function createAppWithZip(projectRootDir,serviceName,projectName,description){
  try{
    _.LOGGER.info(chalk.cyan(i18next.t('init_prj_at'))+ chalk.green(projectRootDir));
    _.LOGGER.info(chalk.cyan(i18next.t('init_installing')));
    var commandString = 'cd \"' + projectRootDir + '\" && npm i --force';
    exec(commandString, function (err) {
      if (err) {
        _.LOGGER.error(chalk.bold.red(err.message));
        return;
      }

      _.LOGGER.info(chalk.cyan(i18next.t('init_projectinitialized'))+ chalk.green(projectRootDir));
      var serviceNo = CONST.SERVICE_NO[serviceName];
      var appTypeName = CONST.SERVICE_APPTYPE_MAPPING[serviceName];
      var appTypeNo = CONST.APPTYPE[appTypeName];
      let tokenObj = store.get('credential', null);
      service.validateTokenObject(tokenObj);
      service.validateDefaultWorkspace(store.get('default_workspace'));
      validate.run(projectRootDir).then(()=>{
        var archivePath = path.join(projectRootDir, '/dist/', path.basename(projectRootDir) + '.zip');
        createApp(archivePath,serviceNo,appTypeNo,projectName,description,projectRootDir);
      });
  });
  }catch(err) {
    _.LOGGER.error(chalk.bold.red(err.message));
  }
}

module.exports = {
  createAppWithoutZip: createAppWithoutZip,
  createAppWithZip: createAppWithZip
};
