// zoho-extension-toolkit/lib/commands/get_workspace.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/

'use strict';


var chalk = require('chalk');
const store = require('../configstore');
const prompt = require('../prompt');
const API = require('../internal/api');
const _ = require('../util_modules');
const Credential = require('../internal/credential');
const font = require('ansi-colors');
var service = require('../utils/service.js');
var i18next = require('i18next');

function run(){
  return new Promise(function (resolve, reject) {
    let tokenObj = store.get('credential', null);
    service.validateTokenObject(tokenObj);
      //for authentication
      Credential.init(tokenObj, false);
      if(store.get('default_workspace')){
          prompt
          .ask(
            prompt.question(
              'collectUsage',
              i18next.t('get_workspace_default')+font.cyan.bold('\''+store.get('default_workspace')+'\'')+'?',
              {
                type: 'confirm',
                defaultAns: true
              }
            )
          )
          .then((ans) => {
            if(!ans.collectUsage){
              getWorkspace().then(()=>{
                resolve();
              }).catch((error) => {
                reject(error);
              });
            }else{
              resolve();
            }
          })
      }else{
        getWorkspace().then(()=>{
          resolve();
        }).catch((error) => {
          reject(error);
        });
      }
    });
  }

  function getWorkspace(){
    return new Promise(function (resolve, reject) {
      _.LOGGER.message(i18next.t('waiting_for_response'))  ;
     //personal app type is workspace_type: '1'
     new API()
     .get('/workspace/restapi/workspace', {
       origin: _.CONSTANTS.ORIGIN.sigma,
       dctype: store.get('dctype'),
       params: {
         workspace_type: 1
           },
       auth: true,
       json: false
     }).then((response) => {
       var orglist = response.body;
      if(orglist.length ==  0){
        _.LOGGER.error(chalk.bold.red(i18next.t('get_workspace_login')+getSourceUrl(_.CONSTANTS.ORIGIN.sigma,  store.get('dctype'))+'.' ));
        reject(i18next.t('get_workspace_no'));
      }else if(orglist.length == 1){
        store.set('default_workspace',orglist[0].org_domain);
        _.LOGGER.success(i18next.t('get_workspace_one_workspace')+chalk.bold.green(font.cyan.bold('\''+store.get('default_workspace')+'\''))+i18next.t('get_workspace_default_success'));
        resolve();
      }else{
         var orglistdata = [];
         orglist.forEach(function(obj) 
         { 
           orglistdata.push({
             'name' : charEncode(obj.org_domain)+' [WorkSpace Owner : '+charEncode(obj.org_admin_displayname)+']',
             'value' : charEncode(obj.org_domain)
           });
         });
         
       prompt
           .ask(
           prompt.question(
               'workspace',
               i18next.t('get_workspace_select'),
               {
               type: 'list',
               choices: orglistdata
               }
           )
           )
           .then((ans) => {
            store.set('default_workspace',ans.workspace);
            _.LOGGER.success(chalk.bold.green( i18next.t('get_workspace_default_set')+font.cyan.bold('\''+store.get('default_workspace')+'\'')));
            resolve();           
          })
          }
        }); 
      });
  }

  function getSourceUrl(url, dctype){
    if(dctype){
			if(_.CONSTANTS.DCTYPES[dctype]){
        url = url.replace('https://', '');
				return url.replace('dctype',_.CONSTANTS.DCTYPES[dctype]);
			}else{
				_.LOGGER.debug("Invalid DC type so using US dc");
				return url.replace('dctype',_.CONSTANTS.DCTYPES["us"]);
			}
		}else{
			_.LOGGER.error(i18next.t('get_workspace_error'));
    }
    return "sigma.zoho.com";
  }

  function charEncode( input ) {
    return String(input)
        .replace(/&lt;/gi, '<')
        .replace(/&gt;/gi, '>')
        .replace(/&#x28;/gi, '(')
        .replace(/&#x29;/gi, ')')
        .replace(/&#x3b;/gi, ';')
        .replace(/&#x2f;/gi, '/');
}

module.exports = {
  run: run
};
