// zoho-extension-toolkit/lib/commands/get_app_files.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/
var chalk = require('chalk');
var path = require('path');
var i18next = require('i18next');
var service = require('../utils/service.js');
const _ = require('../util_modules');
const store = require('../configstore');
const Credential = require('../internal/credential');
var fs = require('fs');
const API = require('../internal/api');
//var unzip = require('unzip');
const fsExtra = require('fs-extra');
var AdmZip = require('adm-zip');

function getAppFiles(app_data,projectRootDir, overwrite){
  return new Promise(function (resolve, reject) {
    let tokenObj = store.get('credential', null);
    service.validateTokenObject(tokenObj);
    Credential.init(tokenObj, false);
    _.LOGGER.message(i18next.t('waiting_for_response'))  ;
    new API()
      .get('/workspace/restapi/'+store.get('default_workspace')+'/apps/internal_ZET_download/'+app_data.appUuid, {
        origin: _.CONSTANTS.ORIGIN.sigma,
        dctype: store.get('dctype'),
        params: {
        
                },
        auth: true,
        json: false,
        downloadfile: true,
        zipPath: projectRootDir
      })
      .then((res) => {
              //  res.pipe(fs.createWriteStream('file.zip'));
              var unzipPath;
              if(overwrite){
                unzipPath = projectRootDir;
              }else{
                unzipPath = projectRootDir+'/temp/cloud';
              }
              deleteAppFile(unzipPath,overwrite).then(()=>{
             // fs.createReadStream(path.join(projectRootDir,'app_file.zip')).pipe(unzip.Extract({ path: projectRootDir }));
             fs.createReadStream(path.join(projectRootDir,'app_file.zip'));
             var am_zip = new AdmZip(path.join(projectRootDir,'app_file.zip'));
             am_zip.extractAllTo(unzipPath, /*overwrite*/true);
             if(overwrite){
              _.LOGGER.success(chalk.bold.green(i18next.t('get_app_success')));
             }
              fs.unlink(path.join(projectRootDir,'app_file.zip'),(err)=>{
                resolve();
              });
            });
              
      });
  });   
}

function deleteAppFile(projectRootDir,overwrite){
  return new Promise(function (resolve, reject) {
    fsExtra.remove(path.join(projectRootDir,'app'),(err)=>{
      if(err){
        _.LOGGER.error(err);
      }
      if (fsExtra.existsSync(path.join(projectRootDir,'plugin-manifest.json'))) {
        fsExtra.removeSync(path.join(projectRootDir,'plugin-manifest.json'));
      }
      if (fsExtra.existsSync(path.join(projectRootDir,'resources.json'))) {
        fsExtra.removeSync(path.join(projectRootDir,'resources.json'));
      }
      if(overwrite){
        resolve();
      }else{
        setTimeout(function() {
          resolve();
      }, 5000);
      }
    });
  });
}

module.exports = {
  //validate using API call 
  run: function (projectRootDir, overwrite) {
    return new Promise(function (resolve, reject) {
      service.validateDefaultWorkspace(store.get('default_workspace'));
      service.validateAppInfoFile(service.isJsonFilePresent(projectRootDir));
      service.readJsonFromFile(projectRootDir).then((app_data) => {
        service.validateJSONData(app_data,store.get('default_workspace'));
        getAppFiles(app_data,projectRootDir, overwrite).then(()=>{
          resolve();
        });
    });
  });
  }
};
