// zoho-extension-toolkit/lib/commands/init.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/

'use strict';

const { exec } = require('child_process');
var { mkdirSync, copySync, existsSync, unlinkSync } = require('fs-extra');
var path = require('path');
var chalk = require('chalk');
var inquirer = require('inquirer');
var serviceUtil = require('../utils/service');
var patchJson = require('../utils/patchJson');
var serviceNames = serviceUtil.getServiceNames();
var i18next = require('i18next');
var CONST = require('../commands/constants.js');
var service = require('../utils/service.js');
var initAPI = require('./init_api');

const SERVICE_NAME = {};
SERVICE_NAME[CONST.SERVICE.CRM] = 'crm';
SERVICE_NAME[CONST.SERVICE.DESK] = 'desk';
SERVICE_NAME[CONST.SERVICE.CATALYST] = 'catalyst';
SERVICE_NAME[CONST.SERVICE.PROJECTS] = 'projects';
SERVICE_NAME[CONST.SERVICE.QNTRL] = 'qntrl';
SERVICE_NAME[CONST.SERVICE.SHOW] = 'show';
SERVICE_NAME[CONST.SERVICE.SDP] = 'sdp';
SERVICE_NAME[CONST.SERVICE.MAIL] = 'mail';
SERVICE_NAME[CONST.SERVICE.CREATOR] = 'creator';
SERVICE_NAME[CONST.SERVICE.FINANCE] = 'finance';
SERVICE_NAME[CONST.SERVICE.IOT] = 'iot';
SERVICE_NAME[CONST.SERVICE.TEAMINBOX] = 'teaminbox';
SERVICE_NAME[CONST.SERVICE.RECRUIT] = 'recruit';
SERVICE_NAME[CONST.SERVICE.BIGIN] = 'bigin';
SERVICE_NAME[CONST.SERVICE.INVENTORY] = 'inventory';
SERVICE_NAME[CONST.SERVICE.PEOPLE] = 'people';
SERVICE_NAME[CONST.SERVICE.EXPENSE] = 'expense';
SERVICE_NAME[CONST.SERVICE.BILLING] = 'billing';
SERVICE_NAME[CONST.SERVICE.FSM] = 'fsm';
SERVICE_NAME[CONST.SERVICE.LOGS360CLOUD] = 'logs360cloud';
//let tokenObj = store.get('credential', null);

const NONINTERACTIVE_INIT_SUPPORTED_SERVICES = [
  CONST.SERVICE.CRM,
  CONST.SERVICE.FINANCE,
  CONST.SERVICE.INVENTORY,
  CONST.SERVICE.EXPENSE,
  CONST.SERVICE.BILLING
];

var promptQuestions = [
  {
    type: 'list',
    name: 'service_name',
    message: i18next.t('init_select_service_msg'),
    choices: serviceNames
  },
  {
    when: function (response) {
      // if(_.JS.isNull(tokenObj)){
      //   return true;
      // }
      return !service.isZetAPIModelServiceByServiceName(response.service_name);
    },
    type: 'input',
    name: 'project_name',
    message: i18next.t('init_projectname'),
    validate: function (value) {
      var isValidName = value.match(/^[A-Za-z0-9_]+$/);
      if (isValidName) { return true; }

      return i18next.t('init_projectname_notempty');
    }
  },
  {
    when: function (response) {
      //&& !service.isZetAPIModelServiceByServiceName(response.service_name) && _.JS.isNull(tokenObj)
      return response.service_name === 'DESK' && !service.isZetAPIModelServiceByServiceName(response.service_name);
    },
    type: 'list',
    name: 'desk_module_support',
    message: i18next.t('init_module_support'),
    choices: [{ name: 'Yes', value: 'Yes' }, { name: 'No', value: 'No' }]
  }
];

// Only for DESK
var isModuleSupport = false;

function run(projectRootDir, options) {
  var serviceName = options && options.zohoService;
  var projectName = options && options.projectName;
  if (serviceName && projectName) {
    if(NONINTERACTIVE_INIT_SUPPORTED_SERVICES.includes(serviceName)){
      if(projectName.match(/^[A-Za-z0-9_]+$/)){
        createProject(projectRootDir, serviceName, projectName);
      }
      else{
        console.log(chalk.bold.red(i18next.t('init_projectname_notempty')));
      }
    }
    else{
      console.log(chalk.bold.red(i18next.t('init_service_not_supported')));
    }
  }
  else{
    inquirer
      .prompt(promptQuestions)
      .then(function (answers) {
        // && !service.isZetAPIModelServiceByServiceName(answers.service_name) && _.JS.isNull(tokenObj)
        // if (answers.service_name === 'DESK') {
        //   additionalQuestion = new Promise(function (resolve, reject) {
        //     let value = answers.desk_module_support === 'Yes' ? true : false;
        //     isModuleSupport = value;
        //     patchJson(path.join(__dirname, '..', '..', 'apptemplate', 'desk', 'plugin-manifest.json'), 'moduleSupport', value);
        //     resolve();
        //   });

        // } else {
        //   additionalQuestion = Promise.resolve();
        // }
        // additionalQuestion = Promise.resolve();
        // additionalQuestion
        //   .then(() => {
            //new flow
            // && !_.JS.isNull(tokenObj)
            if(service.isZetAPIModelServiceByServiceName(answers.service_name)){
              service.sessionExpireCheck().then(()=>{
                initAPI.run(projectRootDir, answers.service_name);
              });
            }else{
              if(answers.service_name == CONST.SERVICE.DESK){
                isModuleSupport = answers.desk_module_support == 'Yes';
              }
              //old flow
              createProject(
                projectRootDir,
                answers.service_name,
                answers.project_name
              );
            }
          // })
          // .catch(function (err) {
          //   console.log(chalk.bold.red(i18next.t('init_err_occur_msg'), err));
          // });
      })
      .catch(function (err) {
        console.log(chalk.bold.red(i18next.t('init_err_occur_msg'), err));
      });
  }
}
function createProject(projectPath, serviceName, projectName) {
  try {
    var template = SERVICE_NAME[serviceName];
    var projectContext = path.join(projectPath, projectName);

    if (existsSync(projectContext)) {
      throw new Error(i18next.t('init_prjdir_with') + projectName + i18next.t('init_alreadyexists'));
    }

    console.log(chalk.cyan(i18next.t('init_prj_at')), chalk.green(projectContext));
    mkdirSync(projectContext);

    // Copy base template:
    copySync(__dirname + '/../../apptemplate/default', projectContext);

    if (serviceName === CONST.SERVICE.CATALYST || serviceName === CONST.SERVICE.PROJECTS
      || serviceName === CONST.SERVICE.QNTRL ) {
      unlinkSync(path.join(projectContext, 'app/widget.html'));
    }

    // Copy service specific files
    if (serviceName === CONST.SERVICE.DESK && typeof template === 'string') {
      patchJson(path.join(__dirname, '..', '..', 'apptemplate', 'desk', 'plugin-manifest.json'), 'moduleSupport', isModuleSupport);
      copySync(__dirname + '/../../apptemplate/' + template, projectContext, {
        filter: (src) => {
          let isPackingFile = src.includes('prepack.js') || src.includes('webpack.config.js');
          let isPackageJsonFile = src.includes('package-module-support.json') || src.includes('package-vannila-js.json');

          return isPackageJsonFile ? false : isPackingFile && isModuleSupport ? true : isPackingFile ? false : true;
        }
      });

      copySync(__dirname + '/../../apptemplate/' + template + (isModuleSupport ? '/package-module-support.json' : '/package-vannila-js.json'), projectContext + '/package.json');
    } else {
      if (typeof template === 'string') {
        copySync(__dirname + '/../../apptemplate/' + template, projectContext);
      }
    }
    
    console.log(chalk.cyan(i18next.t('init_installing')));
    var commandString = 'cd \"' + projectContext + '\" && npm i --force';
    exec(commandString, function (err) {
      if (err) {
        console.error(chalk.bold.red(err.message));
        return;
      }

      console.log(chalk.cyan(i18next.t('init_projectinitialized')), chalk.green(projectContext));
      console.log(chalk.bold.cyan(i18next.t('init_runcmd')));
      console.log(chalk.bold.white(i18next.t('init_cd')), chalk.green('\''+projectName+'\''));
      console.log(chalk.bold.white(i18next.t('init_run')));

      //updating  plugin-manifest.json
      serviceUtil.setServiceName(projectContext, serviceName);
    });
  
  } catch (err) {
    console.log(chalk.bold.red(err.message));
  }
}

module.exports = {
  run: run
};
