// zoho-extension-toolkit/lib/commands/init_api.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/

'use strict';

const { exec } = require('child_process');
var { mkdirSync, copySync, existsSync } = require('fs-extra');
var path = require('path');
var chalk = require('chalk');
var i18next = require('i18next');
var CONST = require('../commands/constants.js');
var workspace = require('./get_workspace');
var createApp = require('./create_app');
const prompt = require('../prompt');
var get_app_files = require('./get_app_files');
const _ = require('../util_modules');
var inquirer = require('inquirer');
var patchJson = require('../utils/patchJson');

var promptQuestions = [
  {
    type: 'input',
    name: 'project_name',
    message: i18next.t('init_projectname'),
    validate: function (value) {
      var isValidName = value.match(/^[A-Za-z0-9_]+$/);
      if (isValidName) { return true; }

      return i18next.t('init_projectname_notempty');
    }
  },
    {
      type: 'input',
      name: 'description',
      message:  i18next.t('app_description'),
      validate: function (value) {
        var isValidName = value.length >= 25;
        if (isValidName) { return true; }
    
        return  i18next.t('app_description_error');
        }
  }];

function run(projectRootDir,service_name) {
  workspace.run().then(()=>{
    inquirer
    .prompt(promptQuestions)
    .then((answers) => {
      createProject(projectRootDir,service_name,answers.project_name,answers.description);
    });
  }).catch((error) => {
    _.LOGGER.debug('Error: ' + error);
  });
}
function createProject(projectPath, serviceName, projectName,description) {
  // Only for DESK
  var isModuleSupport = false;
  try {
    //var template = SERVICE_NAME[serviceName];
    var projectContext = path.join(projectPath, projectName);
    if (existsSync(projectContext)) {
      //_.LOGGER.error(i18next.t('init_prjdir_with') + projectName + i18next.t('init_alreadyexists'));
      throw new Error(i18next.t('init_prjdir_with') + projectName + i18next.t('init_alreadyexists'));
    }
    mkdirSync(projectContext);
    // Copy base template:
    copySync(__dirname + '/../../apptemplate/default', projectContext);
    //for desk
    // Copy service specific files
    if (serviceName === CONST.SERVICE.DESK ) {
      prompt.ask(
        prompt.question(  
        'desk_module_support',
        i18next.t('init_module_support'),
        {
          type: 'list',
        choices: [{ name: 'Yes', value: 'Yes' }, { name: 'No', value: 'No' }]
        })
      ).then((answers) => {
         let value = answers.desk_module_support === 'Yes' ? true : false;
          isModuleSupport = value;
          var template = 'desk';
          copySync(__dirname + '/../../apptemplate/' + template, projectContext, {
            filter: (src) => {
              let isPackingFile = src.includes('prepack.js') || src.includes('webpack.config.js');
              let isPackageJsonFile = src.includes('package-module-support.json') || src.includes('package-vannila-js.json');
              return isPackageJsonFile ? false : isPackingFile && isModuleSupport ? true : isPackingFile ? false : true;
            }
          });
        copySync(__dirname + '/../../apptemplate/' + template + (isModuleSupport ? '/package-module-support.json' : '/package-vannila-js.json'), projectContext + '/package.json'); 
       
        //Need remove this below comment once cloud editor is opened for all user for desk service
        //create_extension(serviceName,projectContext,projectName,description,projectPath,isModuleSupport);
        
        //new flow start
        patchJson(path.join(projectContext, 'plugin-manifest.json'), 'moduleSupport', isModuleSupport);
        createApp.createAppWithZip(projectContext,serviceName,projectName,description)
        //new flow end

    });
    }else{
      create_extension(serviceName,projectContext,projectName,description,projectPath,isModuleSupport);
    } 

    
  } catch (err) {
    _.LOGGER.error(chalk.bold.red(err.message));
  }
}

function create_extension(serviceName,projectContext,projectName,description,projectPath,isModuleSupport){
  try{
    _.LOGGER.info(chalk.cyan(i18next.t('init_prj_at'))+ chalk.green(projectContext));
    _.LOGGER.info(chalk.cyan(i18next.t('init_installing')));
    var commandString = 'cd \"' + projectContext + '\" && npm i --force';
    exec(commandString, function (err) {
      if (err) {
        _.LOGGER.error(chalk.bold.red(err.message));
        return;
      }

      _.LOGGER.info(chalk.cyan(i18next.t('init_projectinitialized'))+chalk.green(projectContext));
      var serviceNo = CONST.SERVICE_NO[serviceName];
      var appTypeName = CONST.SERVICE_APPTYPE_MAPPING[serviceName];
      var appTypeNo = CONST.APPTYPE[appTypeName];
      createApp.createAppWithoutZip(serviceNo,appTypeNo,projectName,description,projectPath).then(()=>{
        get_app_files.run(path.join(projectPath,projectName),true).then(() =>{
          _.LOGGER.info(chalk.bold.cyan(i18next.t('init_runcmd')));
          _.LOGGER.info(chalk.bold.white(i18next.t('init_cd'))+ chalk.green('\''+projectName+'\''));
          _.LOGGER.info(chalk.bold.white(i18next.t('init_run')));
          if(serviceName === CONST.SERVICE.DESK && isModuleSupport){
            patchJson(path.join(projectContext, 'plugin-manifest.json'), 'moduleSupport', isModuleSupport);
          }
        });  
      });
    });
  } catch (err) {
    _.LOGGER.error(chalk.bold.red(err.message));
  }
}

module.exports = {
  run: run
};
