// zoho-extension-toolkit/lib/commands/pull_app.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/
const { exec } = require('child_process');
var { mkdirSync, copySync, existsSync } = require('fs-extra');
var chalk = require('chalk');
var { existsSync } = require('fs-extra');
var path = require('path');
var i18next = require('i18next');
var service = require('../utils/service.js');
var CONST = require('./constants.js');
const _ = require('../util_modules');
const store = require('../configstore');
const Credential = require('../internal/credential');
const API = require('../internal/api');
const prompt = require('../prompt');
var get_app_files = require('./get_app_files');
var workspace = require('./get_workspace');

function getAppList(service_name,projectRootDir){
    let tokenObj = store.get('credential', null);
    service.validateTokenObject(tokenObj);
    service.validateDefaultWorkspace(store.get('default_workspace'));
    Credential.init(tokenObj, false);
    _.LOGGER.message(i18next.t('waiting_for_response'))  ;
    var service_no = CONST.SERVICE_NO[service_name];
    var filter = {};
    filter.service = new Array();
    filter.service.push(service_no);
    new API()
      .get('/workspace/restapi/'+store.get('default_workspace')+'/apps', {
        origin: _.CONSTANTS.ORIGIN.sigma,
        dctype: store.get('dctype'),
        params: {
            filter: JSON.stringify(filter),
            index : 0,
            range : 100
                },
        auth: true,
        json: false
      })
      .then((res) => {
          if(res.body.appsCount == 0){
            _.LOGGER.error(chalk.red(i18next.t('pull_no_extension_found')+'\''+service_name+'\'. '+i18next.t('pull_no_extension_command'))+ chalk.bold.red('Zet init'));
            process.exit(1);
          }
          var data =res.body.appsList;
          var applistdata = [];
         data.forEach(function(obj) 
         { 
            applistdata.push({
              //+' [Description : '+obj.description+']'
             'name' : obj.appName,
             'value' :  obj
           });
         }); 
         prompt.ask(
         prompt.question(
             'appData',
             i18next.t('pull_select_extension'),
             {
             type: 'list',
             choices: applistdata
             }
         )
         )
         .then((ans) => {
            var values=ans.appData;
            var projectName = values.appName;
            var appUUID = values.appUUID;
            var appId = values.app_id;
            var manifest = JSON.parse(values.manifest);
            createProject(projectRootDir,projectName,appUUID, service_name, manifest, appId);
         });         
      });  
}

function createProject(projectPath,  projectName,appUUID, service_name, manifest, appId) {
    try {
     // var template = SERVICE_NAME[serviceName];
      var projectContext = path.join(projectPath, projectName);
  
      if (existsSync(projectContext)) {
        throw new Error(i18next.t('init_prjdir_with') + projectName + i18next.t('init_alreadyexists'));
      }
  
      _.LOGGER.info(chalk.cyan(i18next.t('init_prj_at'))+ chalk.green(projectContext));
      mkdirSync(projectContext);
  
      // Copy base template:
      copySync(__dirname + '/../../apptemplate/default', projectContext);

      //for desk
    // Copy service specific files
    if (service_name === CONST.SERVICE.DESK ) {
        var  isModuleSupport = manifest.moduleSupport;
        _.LOGGER.debug("moduleSupport in Desk manifest : "+isModuleSupport);
        var template = 'desk';
        copySync(__dirname + '/../../apptemplate/' + template, projectContext, {
          filter: (src) => {
            let isPackingFile = src.includes('prepack.js') || src.includes('webpack.config.js');
            let isPackageJsonFile = src.includes('package-module-support.json') || src.includes('package-vannila-js.json');

            return isPackageJsonFile ? false : isPackingFile && isModuleSupport ? true : isPackingFile ? false : true;
          }
        });

        copySync(__dirname + '/../../apptemplate/' + template + (isModuleSupport ? '/package-module-support.json' : '/package-vannila-js.json'), projectContext + '/package.json');
        create_extension(projectContext,projectName,appUUID,projectPath, appId);
    }else{
      create_extension(projectContext,projectName,appUUID,projectPath, appId);
    } 
     
    } catch (err) {
      _.LOGGER.error(chalk.bold.red(err.message));
    }
  }

  function create_extension(projectContext,projectName,appUUID,projectPath, appId){
      try{
      _.LOGGER.info(chalk.cyan(i18next.t('init_installing')));
      var commandString = 'cd \"' + projectContext + '\" && npm i --force';
      exec(commandString, function (err) {
        if (err) {
          console.error(chalk.bold.red(err.message));
          return;
        }

        _.LOGGER.info(chalk.cyan(i18next.t('init_projectinitialized'))+ chalk.green(projectContext));
        var jsonfile = '{ "workspaceName":"'+store.get('default_workspace')+'","appUuid":"'+appUUID+'" ,"appId":"'+appId+'"}';
        service.writeJsonAsFile(JSON.parse(jsonfile), path.join(projectPath, projectName)).then(()=>{
            get_app_files.run(path.join(projectPath,projectName), true).then(() =>{
                _.LOGGER.info(chalk.bold.cyan(i18next.t('init_runcmd')));
                _.LOGGER.info(chalk.bold.white(i18next.t('init_cd'))+chalk.green('\''+projectName+'\''));
                _.LOGGER.info(chalk.bold.white(i18next.t('init_run'))); 
              });
      });
      });
    } catch (err) {
      _.LOGGER.error(chalk.bold.red(err.message));
    }
  }

module.exports = {
  //validate using API call 
  run: function (projectRootDir) {
    let tokenObj = store.get('credential', null);
    service.validateTokenObject(tokenObj);
    if(service.isJsonFilePresent(projectRootDir)){
      service.readJsonFromFile(projectRootDir).then((app_data) => {
        service.validateJSONData(app_data,store.get('default_workspace'));
        prompt.ask(
          prompt.question(  
          'collectUsage',
          i18next.t('pull_warning'),
          {
          type: 'confirm',
          defaultAns: true
          })
        ).then((answers) => {
          if(answers.collectUsage){
            get_app_files.run(projectRootDir,true);
          }
        });
      });
    }else{
      workspace.run().then(()=>{
        prompt.ask(
        prompt.question(
            'service_name',
            i18next.t('pull_select_service'),
            {
            type: 'list',
            choices: service.getZetAPIModelServiceNames
            }
        )
        )
        .then((ans) => {
            getAppList(ans.service_name,projectRootDir);
        });
    }).catch((error) => {
      _.LOGGER.debug('Error: ' + error);
    });
  }
}
};
