// zoho-extension-toolkit/lib/commands/status.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/

var chalk = require('chalk');
var dircompare = require('dir-compare');
var service =require('../utils/service')
var get_app_files = require('./get_app_files'); 
var AdmZip = require('adm-zip');
var path = require('path');
const _ = require('../util_modules');
var fs = require('fs-extra');
const i18next = require('i18next');

module.exports = {
  compare: function(projectRootDir, runCall){
    return new Promise(function (resolve) {
      get_app_files.run(projectRootDir , false).then(() => {
          var unzipPath = projectRootDir+'/temp/local'
          var am_zip = new AdmZip(path.join(projectRootDir, '/dist/', path.basename(projectRootDir) + '.zip'));
          am_zip.extractAllTo(unzipPath, /*overwrite*/true);
          _.LOGGER.debug(chalk.bold.green(i18next.t('status_unzip_success')));
          var options = {
              compareContent: true,
              compareFileSync: dircompare.fileCompareHandlers.lineBasedFileCompare.compareSync,
              compareFileAsync: dircompare.fileCompareHandlers.lineBasedFileCompare.compareAsync,
              ignoreLineEnding: true,
              ignoreWhiteSpaces: true
            };
            
            var path1 = projectRootDir+'/temp/local';
            var path2 = projectRootDir+'/temp/cloud';
            var response = dircompare.compareSync(path1, path2, options);
            //console.log(response.diffSet);
            var objectKeysArray = response.diffSet;
            //console.log(objectKeysArray);
            var data = new Object();
            data.modified = [];
            data.added = [];
            data.deleted = [];
            objectKeysArray.forEach(function(objKey) {
              //var data = new Object();
              //to avoid temp files and not chaged files
              if (( objKey.name1 && objKey.name1.indexOf('.') === 0) || (objKey.name2 && objKey.name2.indexOf('.') === 0) || objKey.state == 'equal') {
                return;
              }else if(objKey.state == 'distinct' && objKey.reason == 'different-content'){
                //For json type we need to check the key value pair is equal or not.
                if(service.getPath(objKey).endsWith(".json")){
                  var jsonFile1 = require(path1+objKey.relativePath+objKey.name1);
                  var jsonFile2 = require(path2+objKey.relativePath+objKey.name2);
                  if(!_.JS.isEqual(jsonFile1,jsonFile2)){
                    data.modified.push(service.getPath(objKey));
                  }
                }else{
                  data.modified.push(service.getPath(objKey));
                }
              }else if(objKey.type1 == 'missing'){
                data.deleted.push(service.getPath(objKey));
              }else if(objKey.type2 == 'missing'){
                data.added.push(service.getPath(objKey));
              }else{
                return;
              }
            });
            _.LOGGER.debug('Status data');
            _.LOGGER.debug(JSON.stringify(data));
            if(data.modified.length == 0 && data.deleted.length == 0 && data.added.length == 0){
              fs.removeSync(path.join(projectRootDir,'temp'));
              if(runCall){
                _.LOGGER.success(chalk.bold.green(i18next.t('status_synced_run')));
              }else{
                _.LOGGER.error(chalk.bold.red(i18next.t('status_synced')));
                process.exit(1);
              }
            }else{
              _.LOGGER.message(chalk.bold.blue(i18next.t('status_not_synced')));
                if(data.modified.length != 0){
                  _.LOGGER.info('');
                  _.LOGGER.warning(chalk.bold.yellow(i18next.t('status_modified_files')));
                  data.modified.forEach(function(val){
                    _.LOGGER.info(chalk.yellow(val));
                  });
                }

                if(data.added.length != 0){
                  _.LOGGER.info('');
                  _.LOGGER.success(chalk.bold.green(i18next.t('status_added_files')));
                  data.added.forEach(function(val){
                    _.LOGGER.info(chalk.green(val));
                  });
                }

                if(data.deleted.length != 0){
                  _.LOGGER.info('');
                  _.LOGGER.error(chalk.bold.red(i18next.t('status_missing_files')));
                  data.deleted.forEach(function(val){
                    _.LOGGER.info(chalk.red(val));
                  });
                }
                fs.removeSync(path.join(projectRootDir,'temp'));

                if(runCall){
                  _.LOGGER.error(chalk.bold.red(i18next.t('status_error')));
                  process.exit(1);
                }
            }
           // console.log(comparsionData);
           resolve();
      });
  });
      
  }
};
