// zoho-extension-toolkit/lib/commands/update_app.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/

var chalk = require('chalk');
var { existsSync, ensureFileSync, createWriteStream, removeSync } = require('fs-extra');
var path = require('path');
var i18next = require('i18next');
var service = require('../utils/service.js');
const _ = require('../util_modules');
const store = require('../configstore');
const Credential = require('../internal/credential');
var fs = require('fs');
const API = require('../internal/api');
const prompt = require('../prompt');
const validate = require('./validation_api');
const status = require('./status');
const FormData = require('form-data');

function updateApp(zipPath,app_data){
  let tokenObj = store.get('credential', null);
  service.validateTokenObject(tokenObj);
  service.validateDefaultWorkspace(store.get('default_workspace'));
  Credential.init(tokenObj, false);
  const form = new FormData();
  form.append('app_zip', fs.createReadStream(zipPath), { filename: 'appZip.zip' } );
  _.LOGGER.message(i18next.t('waiting_for_response'))  ;
   new API()
    .post('/workspace/restapi/'+store.get('default_workspace')+'/apps/'+app_data.appUuid, {
      origin: _.CONSTANTS.ORIGIN.sigma,
      dctype: store.get('dctype'),
      params: {},
      data: form,
      auth: true,
      json: false
    })
    .then((res) => {
              var data = res.body;
              if(data.appObj){
                _.LOGGER.success(chalk.bold.green(i18next.t('update_success')));
              }else{
                _.LOGGER.error(chalk.red(i18next.t('update_failed')));
              }
              });
}

module.exports = {
  //validate using API call 
  run: function (projectRootDir) {
    let tokenObj = store.get('credential', null);
    service.validateTokenObject(tokenObj);
    service.validateDefaultWorkspace(store.get('default_workspace'));
    service.validateAppInfoFile(service.isJsonFilePresent(projectRootDir));
    service.readJsonFromFile(projectRootDir).then((app_data) => {
        service.validateJSONData(app_data,store.get('default_workspace'));
        validate.run(projectRootDir).then(()=>{
          //To compare file changes
          _.LOGGER.success(chalk.bold.green(i18next.t('update_compare')));
          status.compare(projectRootDir, false).then(()=>{
                prompt.ask(
                  prompt.question(  
                  'collectUsage',
                  i18next.t('update_message'),
                  {
                  type: 'confirm',
                  defaultAns: true
                  })
              ).then((answers) => {
                var archivePath = path.join(projectRootDir, '/dist/', path.basename(projectRootDir) + '.zip');
                  if(answers.collectUsage){ 
                      updateApp(archivePath,app_data);
                  }
              });
          });
          
        });
    });
  }
};
