// zoho-extension-toolkit/lib/commands/validation_api.js
/*
Copyright (c) 2017, ZOHO CORPORATION
License: MIT
*/

var archiver = require('archiver');
var chalk = require('chalk');
var { existsSync, ensureFileSync, createWriteStream } = require('fs-extra');
var path = require('path');
var i18next = require('i18next');
var fileTreeParser = require('../utils/file-tree-parser.js');
var service = require('../utils/service.js');
var CONST = require('./constants.js');
const _ = require('../util_modules');
const store = require('../configstore');
const Credential = require('../internal/credential');
var fs = require('fs');
const API = require('../internal/api');
const FormData = require('form-data');

const SERVICE_NO = {};

SERVICE_NO[CONST.SERVICE.CRM] = 2;
SERVICE_NO[CONST.SERVICE.DESK] = 3;
SERVICE_NO[CONST.SERVICE.CATALYST] = 10;
SERVICE_NO[CONST.SERVICE.PROJECTS] = 4;
SERVICE_NO[CONST.SERVICE.QNTRL] = 5;
SERVICE_NO[CONST.SERVICE.SHOW] = 7;
SERVICE_NO[CONST.SERVICE.SDP] = 8;
SERVICE_NO[CONST.SERVICE.MAIL] = 6;
SERVICE_NO[CONST.SERVICE.CREATOR] = 16;
SERVICE_NO[CONST.SERVICE.FINANCE] = 11;
SERVICE_NO[CONST.SERVICE.IOT] = 11;

function bundleFiles(projectContext, archiveStream, fileTree, outputStream) {
  return new Promise(function (resolve) {
    //add files to be zipped
    for (var key in fileTree) {

      if (['projectRoot', 'appBase', 'misc'].indexOf(key) !== -1) continue;

      var assets = fileTree[key];
      for (var i = 0; i < assets.length; i++) {
        var filePath = path.join(fileTree.projectRoot, assets[i]);
        archiveStream.file(filePath, { name: assets[i] });
      }
    }

    var serviceName = service.getServiceName(projectContext);
    if (serviceName !== CONST.SERVICE.CRM && serviceName !== CONST.SERVICE.CREATOR) {
      archiveStream.file(path.join(projectContext, CONST.PLUGIN_MANIFEST), { name: CONST.PLUGIN_MANIFEST });
    }
    if( serviceName === CONST.SERVICE.DESK) {
      var resourceJSONPath = path.join(projectContext, CONST.RESOURCE_JSON);
      if( existsSync(resourceJSONPath) ) {
        archiveStream.file(resourceJSONPath, { name: CONST.RESOURCE_JSON });
      }
    }

    archiveStream.finalize();
    outputStream.on('close', function () {
      resolve();
    });
  });
}

function validateZip(zipPath,service_no,apptype){
  return new Promise(function (resolve) {
  let tokenObj = store.get('credential', null);
  service.validateTokenObject(tokenObj);
  service.validateDefaultWorkspace(store.get('default_workspace'));
  Credential.init(tokenObj, false);
  const form = new FormData();
  form.append('appZip', fs.createReadStream(zipPath), { filename: 'appZip.zip' } );
  _.LOGGER.message(i18next.t('waiting_for_response'))  ;
   new API()
    .post('/workspace/restapi/'+store.get('default_workspace')+'/apps/internal_validate_zip', {
      origin: _.CONSTANTS.ORIGIN.sigma,
      dctype: store.get('dctype'),
      params: {
                  serviceNo: service_no,
                  appTypeNo: apptype
              },
      data: form,
      auth: true,
      json: false
    })
    .then((res) => {
              var data = res.body;
              if(!data.status){
                  fs.unlink( zipPath , function (err) {
                    if (err) {
                      _.LOGGER.debug('Error in deteting the error zip file:'+err);
                    }else{
                      _.LOGGER.debug('Error Zip File deleted!');
                    }
                  }); 
                  var objectKeysArray = Object.keys(data.validationErrors);
                  objectKeysArray.forEach(function(objKey) {
                       var objValue = data.validationErrors[objKey];
                       _.LOGGER.info(chalk.bold.red('\n'+objKey+ ' :- \n'));
                       objValue.forEach(function(element){
                        _.LOGGER.info(chalk.red(' ==> '+ element));
                       });
                  });
                 process.exit(1);
              }else{
                _.LOGGER.success(chalk.bold.green(i18next.t("index_validate_sucess")));
               // _.LOGGER.success(chalk.bold.green(i18next.t('package_sucess') + path.basename(zipPath) + i18next.t('package_in_folder')));
                resolve();
              }
              });
            });
}

function runRule(ruleName, projectContext) {
 
  if (ruleName !== undefined) {
    var ruleFullPath = path.join(__dirname, '../rules', ruleName);
    if (existsSync(ruleFullPath)) {
      return require(ruleFullPath).run(projectContext);
    }
  }
  return Promise.resolve();
}

module.exports = {
  //validate using API call 
  run: function (projectRootDir) {
    return new Promise(function (resolve) {
      let tokenObj = store.get('credential', null);
      service.validateTokenObject(tokenObj);
      var archive = archiver('zip');
      var serviceName = service.getServiceName(projectRootDir);
      //currently the pre rule is only for desk

      //require('./validate').run(projectRootDir, ruleConfig.PRE_RULES);
      var prePackRule = service.getPrePackRule(projectRootDir);

      var archivePath = path.join(projectRootDir, '/dist/', path.basename(projectRootDir) + '.zip');
      ensureFileSync(archivePath);
      var outputStream = createWriteStream(archivePath);
      archive.pipe(outputStream);
      runRule(prePackRule, projectRootDir).then(() => {
        _.LOGGER.message(chalk.bold.cyan(i18next.t('pacakage_prj')));
        var fileTree = fileTreeParser.parse(projectRootDir, serviceName);
        bundleFiles(projectRootDir, archive, fileTree, outputStream).then(() => {
          _.LOGGER.success(chalk.bold.green(i18next.t('package_sucess') + path.basename(archivePath) + i18next.t('package_in_folder')));
          var serviceName = service.getServiceName(projectRootDir);
          var serviceNo = CONST.SERVICE_NO[serviceName];
          var appTypeName = CONST.SERVICE_APPTYPE_MAPPING[serviceName];
          var appTypeNo = CONST.APPTYPE[appTypeName];
          if(serviceNo && appTypeNo){
            validateZip(archivePath,serviceNo,appTypeNo).then(()=>{
              resolve();
            });
          }else{
            _.LOGGER.error(i18next.t('validate_error1')+serviceName+i18next.t('validate_error2'));
            resolve();
          }
        });
      });
    });
  },
  bundleFiles :bundleFiles,
  validateZip :validateZip
};
